// stackpress/cjs/terminal/events/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serve = exports.query = exports.push = exports.purge = exports.migrate = exports.install = exports.generate = exports.emit = exports.drop = exports.build = void 0;
exports.default = listen;
const Server_1 = require("@stackpress/ingest/Server");
const build_js_1 = __importDefault(require("./build.js"));
exports.build = build_js_1.default;
const drop_js_1 = __importDefault(require("./drop.js"));
exports.drop = drop_js_1.default;
const emit_js_1 = __importDefault(require("./emit.js"));
exports.emit = emit_js_1.default;
const generate_js_1 = __importDefault(require("./generate.js"));
exports.generate = generate_js_1.default;
const install_js_1 = __importDefault(require("./install.js"));
exports.install = install_js_1.default;
const migrate_js_1 = __importDefault(require("./migrate.js"));
exports.migrate = migrate_js_1.default;
const purge_js_1 = __importDefault(require("./purge.js"));
exports.purge = purge_js_1.default;
const push_js_1 = __importDefault(require("./push.js"));
exports.push = push_js_1.default;
const query_js_1 = __importDefault(require("./query.js"));
exports.query = query_js_1.default;
const serve_js_1 = __importDefault(require("./serve.js"));
exports.serve = serve_js_1.default;
function listen() {
    const emitter = (0, Server_1.server)();
    emitter.on('build', build_js_1.default);
    emitter.on('drop', drop_js_1.default);
    emitter.on('emit', emit_js_1.default);
    emitter.on('generate', generate_js_1.default);
    emitter.on('install', install_js_1.default);
    emitter.on('migrate', migrate_js_1.default);
    emitter.on('purge', purge_js_1.default);
    emitter.on('push', push_js_1.default);
    emitter.on('query', query_js_1.default);
    emitter.on('serve', serve_js_1.default);
    return emitter;
}
;
