// stackpress/cjs/admin/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminDetailSearchPageFactory = exports.AdminDetailCreatePageFactory = exports.AdminUpdatePageFactory = exports.AdminSearchPageFactory = exports.AdminRestorePageFactory = exports.AdminRemovePageFactory = exports.AdminImportPageFactory = exports.AdminExportPageFactory = exports.AdminDetailPageFactory = exports.AdminCreatePageFactory = void 0;
const create_js_1 = __importDefault(require("./pages/create.js"));
exports.AdminCreatePageFactory = create_js_1.default;
const detail_js_1 = __importDefault(require("./pages/detail.js"));
exports.AdminDetailPageFactory = detail_js_1.default;
const export_js_1 = __importDefault(require("./pages/export.js"));
exports.AdminExportPageFactory = export_js_1.default;
const import_js_1 = __importDefault(require("./pages/import.js"));
exports.AdminImportPageFactory = import_js_1.default;
const remove_js_1 = __importDefault(require("./pages/remove.js"));
exports.AdminRemovePageFactory = remove_js_1.default;
const restore_js_1 = __importDefault(require("./pages/restore.js"));
exports.AdminRestorePageFactory = restore_js_1.default;
const search_js_1 = __importDefault(require("./pages/search.js"));
exports.AdminSearchPageFactory = search_js_1.default;
const update_js_1 = __importDefault(require("./pages/update.js"));
exports.AdminUpdatePageFactory = update_js_1.default;
const create_js_2 = __importDefault(require("./pages/detail/create.js"));
exports.AdminDetailCreatePageFactory = create_js_2.default;
const search_js_2 = __importDefault(require("./pages/detail/search.js"));
exports.AdminDetailSearchPageFactory = search_js_2.default;
