// @stackpress/lib/esm/router/Session.js
import ReadonlyMap from '../data/ReadonlyMap.js';
export class ReadSession extends ReadonlyMap {
    get data() {
        return Object.fromEntries(this._map);
    }
}
export class WriteSession extends ReadSession {
    revisions = new Map();
    clear() {
        for (const name of this.keys()) {
            this.revisions.set(name, { action: 'remove' });
        }
        this._map.clear();
    }
    delete(name) {
        this.revisions.set(name, { action: 'remove' });
        return this._map.delete(name);
    }
    set(name, value) {
        this.revisions.set(name, { action: 'set', value });
        return this._map.set(name, value);
    }
}
export function session(data) {
    const store = new WriteSession(data);
    const callable = Object.assign((name) => store.get(name), {
        clear: () => store.clear(),
        delete: (name) => store.delete(name),
        entries: () => store.entries(),
        forEach: (callback) => store.forEach(callback),
        get: (name) => store.get(name),
        has: (name) => store.has(name),
        keys: () => store.keys(),
        set: (name, value) => store.set(name, value),
        values: () => store.values()
    });
    Object.defineProperty(callable, 'size', { get: () => store.size });
    Object.defineProperty(callable, 'data', { get: () => store.data });
    Object.defineProperty(callable, 'revisions', { get: () => store.revisions });
    return callable;
}
