// frui/cjs/field/Checklist.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useChecklistItem = useChecklistItem;
exports.ChecklistItem = ChecklistItem;
exports.default = Checklist;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_2 = require("react");
function useChecklistItem(config) {
    const { onChange, checked: propChecked, defaultChecked, value, onUpdate } = config;
    const [isChecked, setIsChecked] = (0, react_1.useState)(Boolean(defaultChecked || propChecked));
    (0, react_1.useEffect)(() => {
        if (typeof propChecked === 'undefined')
            return;
        if (propChecked !== isChecked) {
            setIsChecked(propChecked);
        }
    }, [propChecked]);
    return {
        isChecked,
        handlers: {
            change: (e) => {
                setIsChecked(e.target.checked);
                onChange && onChange(e);
                onUpdate && onUpdate(value, e.target.checked);
            },
        },
    };
}
function ChecklistItem(_a) {
    var { label, value, checked: propChecked, defaultChecked, onChange, onUpdate, size = 17, error, disabled, color } = _a, rest = __rest(_a, ["label", "value", "checked", "defaultChecked", "onChange", "onUpdate", "size", "error", "disabled", "color"]);
    const { isChecked, handlers } = useChecklistItem({
        onChange: (e) => onChange && onChange(value, e.target.checked),
        onUpdate,
        checked: propChecked,
        defaultChecked,
        value,
    });
    return ((0, jsx_runtime_1.jsxs)("label", { style: {
            display: 'flex',
            alignItems: 'center',
            gap: '6px',
            fontSize: '15px',
            cursor: disabled ? 'not-allowed' : 'pointer',
            marginTop: '2px',
            color: error ? 'red' : disabled ? 'gray' : 'inherit',
            opacity: disabled ? 0.5 : 1,
        }, children: [(0, jsx_runtime_1.jsx)("input", Object.assign({ type: "checkbox", value: value, checked: isChecked, onChange: handlers.change, style: {
                    width: `${size}px`,
                    height: `${size}px`,
                    cursor: disabled ? 'not-allowed' : 'pointer',
                    verticalAlign: 'middle',
                    marginTop: '2px',
                    accentColor: error ? 'red' : color || 'inherit',
                }, disabled: disabled }, rest)), (0, jsx_runtime_1.jsx)("span", { style: { fontSize: 'inherit', lineHeight: `${size}px` }, children: label })] }));
}
function Checklist({ children, name, defaultValue = [], onChange, onUpdate, style, className, orientation = 'row', size = 17, error, disabled, color, }) {
    const [selected, setSelected] = (0, react_1.useState)(() => {
        const defaultCheckedValues = [];
        react_2.Children.forEach(children, (child) => {
            if ((0, react_2.isValidElement)(child) && child.props.defaultChecked) {
                defaultCheckedValues.push(child.props.value);
            }
        });
        if (defaultValue.length > 0) {
            return defaultValue;
        }
        return defaultCheckedValues;
    });
    const toggleSelection = (value, checked, itemOnUpdate) => {
        if (!disabled) {
            setSelected((prev) => {
                const newSelection = checked
                    ? [...prev, value]
                    : prev.filter((v) => v !== value);
                onChange === null || onChange === void 0 ? void 0 : onChange(newSelection);
                return newSelection;
            });
            setTimeout(() => {
                onUpdate === null || onUpdate === void 0 ? void 0 : onUpdate(value, checked);
                itemOnUpdate === null || itemOnUpdate === void 0 ? void 0 : itemOnUpdate(value, checked);
            }, 0);
        }
    };
    return ((0, jsx_runtime_1.jsxs)("div", { className: `checklist ${className || ''}`, style: Object.assign({ display: 'flex', flexDirection: orientation === 'column' ? 'column' : 'row', alignItems: 'center', gap: '12px', fontSize: '15px', marginTop: '2px', opacity: disabled ? 0.5 : 1 }, style), children: [react_2.Children.map(children, (child) => {
                if ((0, react_2.isValidElement)(child)) {
                    return (0, react_2.cloneElement)(child, {
                        checked: selected.includes(child.props.value),
                        onChange: (value, checked) => {
                            toggleSelection(value, checked, child.props.onUpdate);
                        },
                        size: child.props.size || size,
                        error: child.props.error !== undefined ? child.props.error : error,
                        disabled: disabled || child.props.disabled,
                        color: child.props.color || color,
                        onUpdate: child.props.onUpdate,
                    });
                }
                return child;
            }), (0, jsx_runtime_1.jsx)("input", { type: "hidden", name: name, value: JSON.stringify(selected) })] }));
}
