// frui/cjs/field/Select.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSelect = useSelect;
exports.SelectDropdown = SelectDropdown;
exports.default = Select;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const Input_js_1 = __importDefault(require("./Input.js"));
function useSelect(config) {
    const { value, defaultValue, defaultOptions = [], onDropdown, onSelected, onUpdate, onQuery } = config;
    const currentValue = typeof defaultValue !== 'undefined'
        ? defaultValue
        : typeof value !== 'undefined'
            ? value
            : undefined;
    const defaultSelected = defaultOptions.find(option => option.value === currentValue);
    const [options, setOptions] = (0, react_1.useState)(defaultOptions);
    const [query, setQuery] = (0, react_1.useState)('');
    const [selected, setSelected] = (0, react_1.useState)(defaultSelected);
    const [showing, show] = (0, react_1.useState)(false);
    const handlers = {
        toggle() {
            show(!showing);
            onDropdown && onDropdown(!showing);
        },
        search(e) {
            setTimeout(() => {
                const input = e.target;
                setQuery(input.value);
                onQuery && onQuery(input.value, setOptions);
            });
        },
        match(option) {
            const keyword = (query || '').toLowerCase();
            if (typeof option.keyword === 'string') {
                return option.keyword
                    .toLowerCase()
                    .indexOf(keyword) >= 0;
            }
            else if (typeof option.keyword === 'function') {
                return option.keyword(keyword);
            }
            else if (typeof option.value !== 'undefined'
                && option.value !== null) {
                return option.value
                    .toString()
                    .toLowerCase()
                    .indexOf(keyword) >= 0;
            }
            else if (typeof option.label === 'string') {
                return option.label
                    .toLowerCase()
                    .indexOf(keyword) >= 0;
            }
            return true;
        },
        select(option) {
            show(false);
            setSelected(option);
            onDropdown && onDropdown(false);
            onSelected && onSelected(option);
            onUpdate && onUpdate(option.value);
        }
    };
    (0, react_1.useEffect)(() => {
        if (!value)
            return;
        const selected = options.find(option => option.value === value);
        setSelected(selected);
    }, [value]);
    return { selected, options, showing, handlers };
}
;
function SelectDropdown(props) {
    const { options, show, searchable, select, search, match } = props;
    const style = !show ? { display: 'none' } : undefined;
    return ((0, jsx_runtime_1.jsxs)("div", { className: "frui-field-select-dropdown", style: style, children: [searchable && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-field-select-search", children: [(0, jsx_runtime_1.jsx)(Input_js_1.default, { className: "frui-field-select-search-control", onKeyUp: search }), (0, jsx_runtime_1.jsx)("span", { className: "frui-field-select-search-icon", children: "\uD83D\uDD0E" })] })), (0, jsx_runtime_1.jsx)("div", { className: "frui-field-select-options", children: options.filter(match).map((option, i) => ((0, jsx_runtime_1.jsx)("div", { onClick: _ => select(option), className: "frui-field-select-option", children: option.label }, i))) })] }));
}
;
function Select(props) {
    const { name, searchable, value, defaultValue, options: defaultOptions, placeholder = 'Choose an Option', error, className, style, onQuery, onDropdown, onSelected, onUpdate } = props;
    const entries = (typeof defaultOptions === 'object' && !Array.isArray(defaultOptions)) ? Object.keys(defaultOptions).map(value => ({
        value, label: defaultOptions[value]
    })) : defaultOptions;
    const { selected, options, showing, handlers } = useSelect({
        value,
        defaultValue,
        defaultOptions: entries,
        onQuery,
        onDropdown,
        onSelected,
        onUpdate
    });
    const classNames = ['frui-field-select'];
    if (className) {
        classNames.push(className);
    }
    const placeholderClass = ['frui-field-select-placeholder'];
    if (error) {
        placeholderClass.push('frui-tx-error', 'frui-bd-error');
    }
    const inputValue = typeof (selected === null || selected === void 0 ? void 0 : selected.value) === 'string'
        ? selected.value
        : typeof (selected === null || selected === void 0 ? void 0 : selected.value) === 'number'
            ? String(selected.value)
            : typeof (selected === null || selected === void 0 ? void 0 : selected.value) === 'boolean'
                ? String(selected.value)
                : (selected === null || selected === void 0 ? void 0 : selected.value) && typeof selected.value === 'object'
                    ? JSON.stringify(selected.value)
                    : '';
    return ((0, jsx_runtime_1.jsxs)("div", { className: classNames.join(' '), style: style, children: [(0, jsx_runtime_1.jsx)("div", { className: "frui-field-select-control", onClick: handlers.toggle, children: (selected === null || selected === void 0 ? void 0 : selected.label) || ((0, jsx_runtime_1.jsx)("span", { className: placeholderClass.join(' '), children: placeholder })) }), (0, jsx_runtime_1.jsx)(SelectDropdown, { options: options, show: showing, error: error, searchable: searchable, select: handlers.select, search: handlers.search, match: handlers.match }), (0, jsx_runtime_1.jsx)("input", { name: name, type: "hidden", value: inputValue })] }));
}
;
