// frui/cjs/field/WYSIWYG.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useWYSIWYG = useWYSIWYG;
exports.default = WYSIWYG;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
function useWYSIWYG(config) {
    const { value = '', onChange, onUpdate } = config;
    const [isCodeView, setIsCodeView] = (0, react_1.useState)(false);
    const [currentValue, setCurrentValue] = (0, react_1.useState)(value);
    (0, react_1.useEffect)(() => {
        if (refs.editor.current
            && typeof value === 'string'
            && value !== refs.editor.current.innerHTML
            && !isCodeView) {
            refs.editor.current.innerHTML = value;
            setCurrentValue(value);
            if (refs.hidden.current) {
                refs.hidden.current.value = value;
            }
        }
    }, [value, isCodeView]);
    const refs = {
        editor: (0, react_1.useRef)(null),
        textarea: (0, react_1.useRef)(null),
        hidden: (0, react_1.useRef)(null),
        file: (0, react_1.useRef)(null),
        gallery: (0, react_1.useRef)(null)
    };
    const blocks = {
        listOrdered: () => handlers.execCommand('insertOrderedList'),
        listUnordered: () => handlers.execCommand('insertUnorderedList'),
        undo: () => handlers.execCommand('undo'),
        redo: () => handlers.execCommand('redo'),
        font(e) {
            handlers.execCommand('fontName', e.target.value);
            e.target.value = '';
        },
        size(e) {
            handlers.execCommand('fontSize', e.target.value);
            e.target.value = '';
        },
        format(e) {
            handlers.execCommand('formatBlock', e.target.value);
        },
        paragraph: () => handlers.execCommand('formatBlock', 'p'),
        blockquote: () => handlers.execCommand('formatBlock', 'blockquote'),
        bold: () => handlers.execCommand('bold'),
        italic: () => handlers.execCommand('italic'),
        underline: () => handlers.execCommand('underline'),
        color(e) {
            handlers.execCommand('foreColor', e.target.value);
        },
        highlight(e) {
            handlers.execCommand('hiliteColor', e.target.value);
        },
        strikethrough: () => handlers.execCommand('strikeThrough'),
        subscript: () => handlers.execCommand('subscript'),
        superscript: () => handlers.execCommand('superscript'),
        textStyle(e) {
            const style = e.target.value;
            if (!refs.editor.current || isCodeView)
                return;
            const selection = window.getSelection();
            if (selection && selection.rangeCount > 0) {
                const range = selection.getRangeAt(0);
                const fragment = range.extractContents();
                const styleSpans = fragment.querySelectorAll('span[class*="__frui-wysiwyg-t-"]');
                styleSpans.forEach(span => {
                    const parent = span.parentNode;
                    while (span.firstChild) {
                        parent === null || parent === void 0 ? void 0 : parent.insertBefore(span.firstChild, span);
                    }
                    parent === null || parent === void 0 ? void 0 : parent.removeChild(span);
                });
                const span = document.createElement('span');
                span.className = `__frui-wysiwyg-t-${style}`;
                span.appendChild(fragment);
                range.insertNode(span);
                selection.removeAllRanges();
                const newRange = document.createRange();
                newRange.selectNodeContents(span);
                selection.addRange(newRange);
            }
            handlers.input();
            e.target.value = '';
        },
        removeFormat: () => {
            if (!refs.editor.current || isCodeView)
                return;
            const selection = window.getSelection();
            if (selection && selection.rangeCount > 0) {
                const range = selection.getRangeAt(0);
                const container = document.createElement('div');
                container.appendChild(range.cloneContents());
                const styledElements = container.querySelectorAll('b, i, u, strike, sub, sup, font, span[style], '
                    + 'span[class*="__frui-wysiwyg-t-"]');
                styledElements.forEach(element => {
                    var _a;
                    const textNode = document.createTextNode(element.textContent || '');
                    (_a = element.parentNode) === null || _a === void 0 ? void 0 : _a.replaceChild(textNode, element);
                });
                container.normalize();
                const p = document.createElement('p');
                p.innerHTML = container.innerHTML || ' ';
                range.deleteContents();
                range.insertNode(p);
                selection.removeAllRanges();
                selection.addRange(range);
            }
            handlers.input();
            onUpdate && onUpdate({
                value: refs.editor.current.innerHTML || '',
                action: 'removeFormat'
            });
        },
        indent: () => handlers.execCommand('indent'),
        outdent: () => handlers.execCommand('outdent'),
        alignLeft: () => handlers.execCommand('justifyLeft'),
        alignCenter: () => handlers.execCommand('justifyCenter'),
        alignRight: () => handlers.execCommand('justifyRight'),
        rule: () => handlers.execCommand('insertHorizontalRule'),
        lineHeight(e) {
            if (refs.editor.current && !isCodeView) {
                refs.editor.current.style.lineHeight = e.target.value;
                handlers.input();
            }
        },
        table() {
            if (isCodeView)
                return;
            const rows = prompt('Enter number of rows', '2');
            const cols = prompt('Enter number of columns', '2');
            if (rows && cols && !isNaN(+rows) && !isNaN(+cols)) {
                const tableHTML = `<table border="1">${Array.from({ length: parseInt(rows) }, () => `<tr>${Array.from({ length: parseInt(cols) }, () => '<td> </td>').join('')}</tr>`).join('')}</table>`;
                handlers.execCommand('insertHTML', tableHTML);
            }
        },
        link() {
            if (isCodeView)
                return;
            const url = prompt('Enter URL', 'https://');
            if (url && /^https?:\/\//.test(url))
                handlers.execCommand('createLink', url);
        },
        image: () => { var _a; return (_a = refs.file.current) === null || _a === void 0 ? void 0 : _a.click(); },
        imageUpload(e) {
            var _a;
            if (isCodeView)
                return;
            const file = (_a = e.target.files) === null || _a === void 0 ? void 0 : _a[0];
            if (file && file.type.startsWith('image/')) {
                const reader = new FileReader();
                reader.onload = () => handlers.execCommand('insertImage', reader.result);
                reader.readAsDataURL(file);
            }
        },
        imageGallery: () => { var _a; return (_a = refs.gallery.current) === null || _a === void 0 ? void 0 : _a.click(); },
        galleryUpload(e) {
            if (isCodeView)
                return;
            const files = e.target.files;
            if (files) {
                Array.from(files).forEach((file) => {
                    if (file.type.startsWith('image/')) {
                        const reader = new FileReader();
                        reader.onload = () => handlers.execCommand('insertImage', reader.result);
                        reader.readAsDataURL(file);
                    }
                });
            }
        },
        video() {
            if (isCodeView)
                return;
            const url = prompt('Enter video URL (e.g., YouTube embed)', '');
            if (url && /^https?:\/\//.test(url)) {
                handlers.execCommand('insertHTML', `<iframe src="${url}" width="560" height="315" frameborder="0" allowfullscreen></iframe>`);
            }
        },
        audio() {
            if (isCodeView)
                return;
            const url = prompt('Enter audio URL', '');
            if (url && /^https?:\/\//.test(url)) {
                handlers.execCommand('insertHTML', `<audio controls src="${url}"></audio>`);
            }
        },
        math() {
            if (isCodeView)
                return;
            const expr = prompt('Enter math expression (e.g., x^2)', '');
            if (expr) {
                handlers.execCommand('insertHTML', `<span class="frui-wysiwyg-math">[Math: ${expr}]</span>`);
            }
        },
        fullscreen: () => {
            if (refs.editor.current) {
                if (!document.fullscreenElement) {
                    refs.editor.current.requestFullscreen();
                }
                else {
                    document.exitFullscreen();
                }
            }
        },
        showBlocks() {
            if (refs.editor.current && !isCodeView) {
                refs.editor.current.classList.toggle('frui-wysiwyg-show-block');
            }
        },
        codeViewToggle() {
            if (refs.editor.current && refs.textarea.current && refs.hidden.current) {
                if (!isCodeView) {
                    refs.textarea.current.value = refs.editor.current.innerHTML;
                    refs.hidden.current.value = refs.editor.current.innerHTML;
                    refs.textarea.current.style.display = 'block';
                    refs.editor.current.style.display = 'none';
                }
                else {
                    refs.editor.current.innerHTML = refs.textarea.current.value;
                    refs.hidden.current.value = refs.textarea.current.value;
                    refs.textarea.current.style.display = 'none';
                    refs.editor.current.style.display = 'block';
                }
                setIsCodeView(!isCodeView);
                setCurrentValue(!isCodeView ? refs.textarea.current.value : refs.editor.current.innerHTML);
                onChange && onChange(!isCodeView ? refs.textarea.current.value : refs.editor.current.innerHTML);
                onUpdate && onUpdate({ value: !isCodeView ? refs.textarea.current.value : refs.editor.current.innerHTML, action: 'codeViewToggle' });
            }
        },
        preview() {
            if (refs.editor.current) {
                const win = window.open('', '_blank');
                win === null || win === void 0 ? void 0 : win.document.write(`<html><body>${refs.editor.current.innerHTML}</body></html>`);
                win === null || win === void 0 ? void 0 : win.document.close();
            }
        },
        print: () => window.print(),
        save() {
            if (refs.editor.current) {
                const blob = new Blob([refs.editor.current.innerHTML], { type: 'text/html' });
                const url = URL.createObjectURL(blob);
                const a = document.createElement('a');
                a.href = url;
                a.download = 'editor-content.html';
                a.click();
                URL.revokeObjectURL(url);
            }
        },
        template(e) {
            const templateHTML = e.target.value;
            if (templateHTML && !isCodeView)
                handlers.execCommand('insertHTML', templateHTML);
            e.target.value = '';
        },
        dirToggle() {
            if (refs.editor.current && !isCodeView) {
                const currentDir = refs.editor.current.getAttribute('dir') || 'ltr';
                refs.editor.current.setAttribute('dir', currentDir === 'ltr' ? 'rtl' : 'ltr');
                handlers.input();
            }
        },
        dirLTR() {
            if (refs.editor.current && !isCodeView) {
                refs.editor.current.setAttribute('dir', 'ltr');
                handlers.input();
            }
        },
        dirRTL() {
            if (refs.editor.current && !isCodeView) {
                refs.editor.current.setAttribute('dir', 'rtl');
                handlers.input();
            }
        }
    };
    const handlers = {
        change(e) {
            if (isCodeView && refs.hidden.current) {
                refs.hidden.current.value = e.target.value;
                setCurrentValue(e.target.value);
                onChange === null || onChange === void 0 ? void 0 : onChange(e.target.value);
                onUpdate === null || onUpdate === void 0 ? void 0 : onUpdate({ value: e.target.value, action: 'input' });
            }
        },
        input() {
            if (refs.editor.current && refs.hidden.current) {
                const content = refs.editor.current.innerHTML;
                refs.hidden.current.value = content;
                setCurrentValue(content);
                onChange === null || onChange === void 0 ? void 0 : onChange(content);
                onUpdate === null || onUpdate === void 0 ? void 0 : onUpdate({ value: content, action: 'input' });
            }
        },
        click(e) {
            if (isCodeView || !refs.editor.current)
                return;
            const target = e.target;
            const link = target.closest('a[href]');
            if (link && (e.ctrlKey || e.metaKey)) {
                e.preventDefault();
                const href = link.getAttribute('href');
                if (href) {
                    window.open(href, '_blank', 'noopener,noreferrer');
                }
            }
        },
        execCommand(command, value) {
            if (!refs.editor.current || isCodeView)
                return;
            refs.editor.current.focus();
            document.execCommand(command, false, value);
            refs.editor.current.focus();
            handlers.input();
            onUpdate === null || onUpdate === void 0 ? void 0 : onUpdate({ value: refs.editor.current.innerHTML || '', action: command });
        }
    };
    return { refs, blocks, handlers, isCodeView, value: currentValue };
}
;
function WYSIWYG(props) {
    const { history, font, size, format, paragraph, blockquote, style, color, highlight, text, textStyle, remove, indent, align, rule, list, lineheight, table, link, image, imageGallery, video, audio, math, fullscreen, showblocks, code, preview, print, save, template, dir, onChange, onUpdate } = props, attributes = __rest(props, ["history", "font", "size", "format", "paragraph", "blockquote", "style", "color", "highlight", "text", "textStyle", "remove", "indent", "align", "rule", "list", "lineheight", "table", "link", "image", "imageGallery", "video", "audio", "math", "fullscreen", "showblocks", "code", "preview", "print", "save", "template", "dir", "onChange", "onUpdate"]);
    const { refs, blocks, handlers, value, isCodeView } = useWYSIWYG({
        value: props.value,
        onChange,
        onUpdate
    });
    return ((0, jsx_runtime_1.jsxs)("div", { className: `frui-wysiwyg ${dir === 'rtl' ? 'frui-wysiwyg-rtl' : ''}`, children: [(0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-toolbar", children: [history && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.undo, title: "Undo", "aria-label": "Undo", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-undo" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.redo, title: "Redo", "aria-label": "Redo", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-redo" }) })] })), font && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsxs)("select", { className: "frui-wysiwyg-btn-select", onChange: blocks.font, title: "Font", "aria-label": "Font", defaultValue: "", disabled: isCodeView, children: [(0, jsx_runtime_1.jsx)("option", { value: "", disabled: true, children: "Font" }), (0, jsx_runtime_1.jsx)("option", { value: "Arial", style: { fontFamily: 'Arial' }, children: "Arial" }), (0, jsx_runtime_1.jsx)("option", { value: "Comic Sans MS", style: { fontFamily: 'Comic Sans MS' }, children: "Comic Sans MS" }), (0, jsx_runtime_1.jsx)("option", { value: "Courier New", style: { fontFamily: 'Courier New' }, children: "Courier New" }), (0, jsx_runtime_1.jsx)("option", { value: "Impact", style: { fontFamily: 'Impact' }, children: "Impact" }), (0, jsx_runtime_1.jsx)("option", { value: "Georgia", style: { fontFamily: 'Georgia' }, children: "Georgia" }), (0, jsx_runtime_1.jsx)("option", { value: "Tahoma", style: { fontFamily: 'Tahoma' }, children: "Tahoma" }), (0, jsx_runtime_1.jsx)("option", { value: "Trebuchet MS", style: { fontFamily: 'Trebuchet MS' }, children: "Trebuchet MS" }), (0, jsx_runtime_1.jsx)("option", { value: "Verdana", style: { fontFamily: 'Verdana' }, children: "Verdana" })] }) })), size && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsxs)("select", { className: "frui-wysiwyg-btn-select", onChange: blocks.size, title: "Size", "aria-label": "Font Size", defaultValue: "", disabled: isCodeView, children: [(0, jsx_runtime_1.jsx)("option", { value: "", disabled: true, children: "Size" }), (0, jsx_runtime_1.jsx)("option", { value: "8px", children: "8px" }), (0, jsx_runtime_1.jsx)("option", { value: "9px", children: "9px" }), (0, jsx_runtime_1.jsx)("option", { value: "10px", children: "10px" }), (0, jsx_runtime_1.jsx)("option", { value: "11px", children: "11px" }), (0, jsx_runtime_1.jsx)("option", { value: "12px", children: "12px" }), (0, jsx_runtime_1.jsx)("option", { value: "14px", children: "14px" }), (0, jsx_runtime_1.jsx)("option", { value: "16px", children: "16px" }), (0, jsx_runtime_1.jsx)("option", { value: "18px", children: "18px" }), (0, jsx_runtime_1.jsx)("option", { value: "20px", children: "20px" }), (0, jsx_runtime_1.jsx)("option", { value: "22px", children: "22px" }), (0, jsx_runtime_1.jsx)("option", { value: "24px", children: "24px" }), (0, jsx_runtime_1.jsx)("option", { value: "26px", children: "26px" }), (0, jsx_runtime_1.jsx)("option", { value: "28px", children: "28px" }), (0, jsx_runtime_1.jsx)("option", { value: "36px", children: "36px" }), (0, jsx_runtime_1.jsx)("option", { value: "48px", children: "48px" }), (0, jsx_runtime_1.jsx)("option", { value: "72px", children: "72px" })] }) })), format && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsxs)("select", { className: "frui-wysiwyg-btn-select", onChange: blocks.format, title: "Format", "aria-label": "Block Format", defaultValue: "", disabled: isCodeView, children: [(0, jsx_runtime_1.jsx)("option", { value: "", disabled: true, children: "Format" }), (0, jsx_runtime_1.jsx)("option", { value: "h1", children: "Heading 1" }), (0, jsx_runtime_1.jsx)("option", { value: "h2", children: "Heading 2" }), (0, jsx_runtime_1.jsx)("option", { value: "h3", children: "Heading 3" }), (0, jsx_runtime_1.jsx)("option", { value: "h4", children: "Heading 4" }), (0, jsx_runtime_1.jsx)("option", { value: "h5", children: "Heading 5" }), (0, jsx_runtime_1.jsx)("option", { value: "h6", children: "Heading 6" }), (0, jsx_runtime_1.jsx)("option", { value: "p", children: "Paragraph" }), (0, jsx_runtime_1.jsx)("option", { value: "pre", children: "Code" })] }) })), paragraph && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.paragraph, title: "Paragraph", "aria-label": "Paragraph", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-paragraph" }) }) })), blockquote && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.blockquote, title: "Blockquote", "aria-label": "Blockquote", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-quote-right" }) }) })), style && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.bold, title: "Bold", "aria-label": "Bold", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-bold" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.italic, title: "Italic", "aria-label": "Italic", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-italic" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.underline, title: "Underline", "aria-label": "Underline", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-underline" }) })] })), color && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("input", { type: "color", onChange: blocks.color, title: "Text Color", "aria-label": "Text Color", style: { width: '34px', height: '34px', padding: 0, border: 'none' }, disabled: isCodeView }) })), highlight && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("input", { type: "color", onChange: blocks.highlight, title: "Highlight", "aria-label": "Highlight", style: { width: '34px', height: '34px', padding: 0, border: 'none' }, disabled: isCodeView }) })), text && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.strikethrough, title: "Strikethrough", "aria-label": "Strikethrough", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-strikethrough" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.subscript, title: "Subscript", "aria-label": "Subscript", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-subscript" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.superscript, title: "Superscript", "aria-label": "Superscript", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-superscript" }) })] })), textStyle && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsxs)("select", { className: "frui-wysiwyg-btn-select", onChange: blocks.textStyle, title: "Text Style", "aria-label": "Text Style", defaultValue: "", disabled: isCodeView, children: [(0, jsx_runtime_1.jsx)("option", { value: "", disabled: true, children: "T Style" }), (0, jsx_runtime_1.jsx)("option", { value: "code", style: { backgroundColor: '#F9F9F9',
                                        border: '1px solid #E1E1E1', padding: '0 4px' }, children: "Code" }), (0, jsx_runtime_1.jsx)("option", { value: "translucent", style: { opacity: 0.6 }, children: "Translucent" }), (0, jsx_runtime_1.jsx)("option", { value: "shadow", style: { textShadow: '2px 2px 2px #B1B1B1' }, children: "Shadow" }), (0, jsx_runtime_1.jsx)("option", { value: "spaced", style: {
                                        letterSpacing: '2px',
                                        wordSpacing: '4px'
                                    }, children: "Spaced" }), (0, jsx_runtime_1.jsx)("option", { value: "bordered", style: {
                                        border: '2px solid #000000',
                                        padding: '2px 6px'
                                    }, children: "Bordered" }), (0, jsx_runtime_1.jsx)("option", { value: "neon", style: {
                                        textShadow: '0 0 5px #FFFFFF, 0 0 10px #FFFFFF, 0 0 20px #FF00DE',
                                        color: '#FFFFFF',
                                        backgroundColor: '#000000',
                                        padding: '2px 6px'
                                    }, children: "NEON" })] }) })), remove && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.removeFormat, title: "Remove Format", "aria-label": "Remove Format", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-eraser" }) }) })), indent && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.indent, title: "Indent", "aria-label": "Indent", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-indent" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.outdent, title: "Outdent", "aria-label": "Outdent", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-outdent" }) })] })), align && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.alignLeft, title: "Align Left", "aria-label": "Align Left", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-align-left" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.alignCenter, title: "Align Center", "aria-label": "Align Center", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-align-center" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.alignRight, title: "Align Right", "aria-label": "Align Right", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-align-right" }) })] })), rule && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.rule, title: "Horizontal Rule", "aria-label": "Horizontal Rule", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-minus" }) }) })), list && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.listOrdered, title: "Ordered List", "aria-label": "Ordered List", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-list-ol" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.listUnordered, title: "Unordered List", "aria-label": "Unordered List", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-list-ul" }) })] })), lineheight && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsxs)("select", { className: "frui-wysiwyg-btn-select", onChange: blocks.lineHeight, title: "Line Height", "aria-label": "Line Height", defaultValue: "1", disabled: isCodeView, children: [(0, jsx_runtime_1.jsx)("option", { value: "1", children: "1" }), (0, jsx_runtime_1.jsx)("option", { value: "1.15", children: "1.15" }), (0, jsx_runtime_1.jsx)("option", { value: "1.5", children: "1.5" }), (0, jsx_runtime_1.jsx)("option", { value: "2", children: "2" })] }) })), table && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.table, title: "Table", "aria-label": "Insert Table", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-table" }) }) })), link && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.link, title: "Link", "aria-label": "Insert Link", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-link" }) }) })), image && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.image, title: "Image", "aria-label": "Insert Image", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-image" }) }), (0, jsx_runtime_1.jsx)("input", { type: "file", ref: refs.file, onChange: blocks.imageUpload, accept: "image/*", style: { display: 'none' }, "aria-hidden": "true" })] })), imageGallery && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.imageGallery, title: "Image Gallery", "aria-label": "Insert from Gallery", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-images" }) }), (0, jsx_runtime_1.jsx)("input", { type: "file", ref: refs.gallery, onChange: blocks.galleryUpload, accept: "image/*", multiple: true, style: { display: 'none' }, "aria-hidden": "true" })] })), video && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.video, title: "Video", "aria-label": "Insert Video", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-video" }) }) })), audio && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.audio, title: "Audio", "aria-label": "Insert Audio", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-volume-up" }) }) })), math && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.math, title: "Math", "aria-label": "Insert Math", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-square-root-alt" }) }) })), fullscreen && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.fullscreen, title: "Fullscreen", "aria-label": "Toggle Fullscreen", children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-expand" }) }) })), showblocks && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.showBlocks, title: "Show Blocks", "aria-label": "Toggle Block Visibility", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-border-all" }) }) })), code && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.codeViewToggle, title: isCodeView ? 'WYSIWYG View' : 'Code View', "aria-label": isCodeView
                                ? 'Switch to WYSIWYG View'
                                : 'Switch to Code View', children: (0, jsx_runtime_1.jsx)("i", { className: `fas fa-fw ${isCodeView ? 'fa-eye' : 'fa-code'}` }) }) })), preview && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.preview, title: "Preview", "aria-label": "Preview Content", children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-eye" }) }) })), print && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.print, title: "Print", "aria-label": "Print", children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-print" }) }) })), save && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.save, title: "Save", "aria-label": "Save as HTML", children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-save" }) }) })), template && ((0, jsx_runtime_1.jsx)("div", { className: "frui-wysiwyg-btn-module", children: (0, jsx_runtime_1.jsxs)("select", { className: "frui-wysiwyg-btn-select", onChange: blocks.template, title: "Template", "aria-label": "Insert Template", defaultValue: "", disabled: isCodeView, children: [(0, jsx_runtime_1.jsx)("option", { value: "", disabled: true, children: "Template" }), (0, jsx_runtime_1.jsx)("option", { value: "<p><strong>Header</strong><br>Content</p>", children: "Simple Block" }), (0, jsx_runtime_1.jsx)("option", { value: "<table border='1'><tr><td>Cell</td></tr></table>", children: "Table Block" })] }) })), dir && ((0, jsx_runtime_1.jsxs)("div", { className: "frui-wysiwyg-btn-module", children: [(0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.dirToggle, title: "Toggle Direction", "aria-label": "Toggle Text Direction", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-exchange-alt" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.dirLTR, title: "Left to Right", "aria-label": "Left to Right", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-long-arrow-alt-left" }) }), (0, jsx_runtime_1.jsx)("button", { className: "frui-wysiwyg-btn", onClick: blocks.dirRTL, title: "Right to Left", "aria-label": "Right to Left", disabled: isCodeView, children: (0, jsx_runtime_1.jsx)("i", { className: "fas fa-fw fa-long-arrow-alt-right" }) })] }))] }), (0, jsx_runtime_1.jsx)("input", Object.assign({ type: "hidden", ref: refs.hidden, value: value }, attributes)), (0, jsx_runtime_1.jsx)("div", { ref: refs.editor, className: "frui-wysiwyg-editable", contentEditable: !isCodeView, onInput: handlers.input, onClick: handlers.click, dir: dir, "aria-label": "Rich Text Editor" }), (0, jsx_runtime_1.jsx)("textarea", { ref: refs.textarea, style: {
                    display: isCodeView ? 'block' : 'none',
                    width: '100%',
                    minHeight: '200px',
                    padding: '16px',
                    fontFamily: 'monospace',
                    fontSize: '13px',
                    backgroundColor: '#F9F9F9',
                    border: 'none',
                    resize: 'vertical'
                }, onChange: handlers.change, defaultValue: value })] }));
}
;
