// @stackpress/lib/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withUnknownHost = exports.shouldBeAnArray = exports.objectFromFormData = exports.objectFromQuery = exports.objectFromJson = exports.objectFromArgs = exports.makeObject = exports.makeArray = exports.isObject = exports.formDataToObject = exports.nest = exports.set = exports.map = exports.cookie = exports.control = exports.codes = exports.Status = exports.Reflection = exports.Exception = exports.Terminal = exports.Router = exports.Response = exports.Request = exports.RouteEmitter = exports.ExpressEmitter = exports.EventEmitter = exports.TaskQueue = exports.ItemQueue = exports.NodeFS = exports.FileLoader = exports.Nest = exports.ReadonlySet = exports.ReadonlyPath = exports.ReadonlyNest = exports.ReadonlyMap = exports.QueryString = exports.PathString = exports.FormData = exports.FileData = exports.ArgString = void 0;
const ArgString_js_1 = __importDefault(require("./data/processors/ArgString.js"));
exports.ArgString = ArgString_js_1.default;
const FileData_js_1 = __importDefault(require("./data/processors/FileData.js"));
exports.FileData = FileData_js_1.default;
const FormData_js_1 = __importDefault(require("./data/processors/FormData.js"));
exports.FormData = FormData_js_1.default;
const PathString_js_1 = __importDefault(require("./data/processors/PathString.js"));
exports.PathString = PathString_js_1.default;
const QueryString_js_1 = __importDefault(require("./data/processors/QueryString.js"));
exports.QueryString = QueryString_js_1.default;
const ReadonlyMap_js_1 = __importDefault(require("./data/ReadonlyMap.js"));
exports.ReadonlyMap = ReadonlyMap_js_1.default;
const ReadonlyNest_js_1 = __importDefault(require("./data/ReadonlyNest.js"));
exports.ReadonlyNest = ReadonlyNest_js_1.default;
const ReadonlyPath_js_1 = __importDefault(require("./data/ReadonlyPath.js"));
exports.ReadonlyPath = ReadonlyPath_js_1.default;
const ReadonlySet_js_1 = __importDefault(require("./data/ReadonlySet.js"));
exports.ReadonlySet = ReadonlySet_js_1.default;
const Nest_js_1 = __importStar(require("./data/Nest.js"));
exports.Nest = Nest_js_1.default;
Object.defineProperty(exports, "formDataToObject", { enumerable: true, get: function () { return Nest_js_1.formDataToObject; } });
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return Nest_js_1.isObject; } });
Object.defineProperty(exports, "makeArray", { enumerable: true, get: function () { return Nest_js_1.makeArray; } });
Object.defineProperty(exports, "makeObject", { enumerable: true, get: function () { return Nest_js_1.makeObject; } });
Object.defineProperty(exports, "objectFromArgs", { enumerable: true, get: function () { return Nest_js_1.objectFromArgs; } });
Object.defineProperty(exports, "objectFromJson", { enumerable: true, get: function () { return Nest_js_1.objectFromJson; } });
Object.defineProperty(exports, "objectFromQuery", { enumerable: true, get: function () { return Nest_js_1.objectFromQuery; } });
Object.defineProperty(exports, "objectFromFormData", { enumerable: true, get: function () { return Nest_js_1.objectFromFormData; } });
Object.defineProperty(exports, "shouldBeAnArray", { enumerable: true, get: function () { return Nest_js_1.shouldBeAnArray; } });
Object.defineProperty(exports, "nest", { enumerable: true, get: function () { return Nest_js_1.nest; } });
const map_js_1 = __importDefault(require("./data/map.js"));
exports.cookie = map_js_1.default;
const map_js_2 = __importDefault(require("./data/map.js"));
exports.map = map_js_2.default;
const set_js_1 = __importDefault(require("./data/set.js"));
exports.set = set_js_1.default;
const FileLoader_js_1 = __importDefault(require("./system/FileLoader.js"));
exports.FileLoader = FileLoader_js_1.default;
const NodeFS_js_1 = __importDefault(require("./system/NodeFS.js"));
exports.NodeFS = NodeFS_js_1.default;
const ItemQueue_js_1 = __importDefault(require("./queue/ItemQueue.js"));
exports.ItemQueue = ItemQueue_js_1.default;
const TaskQueue_js_1 = __importDefault(require("./queue/TaskQueue.js"));
exports.TaskQueue = TaskQueue_js_1.default;
const EventEmitter_js_1 = __importDefault(require("./emitter/EventEmitter.js"));
exports.EventEmitter = EventEmitter_js_1.default;
const ExpressEmitter_js_1 = __importDefault(require("./emitter/ExpressEmitter.js"));
exports.ExpressEmitter = ExpressEmitter_js_1.default;
const RouteEmitter_js_1 = __importDefault(require("./emitter/RouteEmitter.js"));
exports.RouteEmitter = RouteEmitter_js_1.default;
const Request_js_1 = __importStar(require("./router/Request.js"));
exports.Request = Request_js_1.default;
Object.defineProperty(exports, "withUnknownHost", { enumerable: true, get: function () { return Request_js_1.withUnknownHost; } });
const Response_js_1 = __importDefault(require("./router/Response.js"));
exports.Response = Response_js_1.default;
const Router_js_1 = __importDefault(require("./router/Router.js"));
exports.Router = Router_js_1.default;
const Terminal_js_1 = __importStar(require("./router/Terminal.js"));
exports.Terminal = Terminal_js_1.default;
Object.defineProperty(exports, "control", { enumerable: true, get: function () { return Terminal_js_1.control; } });
const Exception_js_1 = __importDefault(require("./Exception.js"));
exports.Exception = Exception_js_1.default;
const Reflection_js_1 = __importDefault(require("./Reflection.js"));
exports.Reflection = Reflection_js_1.default;
const Status_js_1 = __importStar(require("./Status.js"));
exports.Status = Status_js_1.default;
Object.defineProperty(exports, "codes", { enumerable: true, get: function () { return Status_js_1.codes; } });
