// stackpress/esm/schema/config/typemaps.js
export const type = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'string',
    Time: 'string',
    Datetime: 'string',
    Json: 'string',
    Object: 'string',
    Hash: 'string'
};
export const model = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'string',
    Time: 'string',
    Datetime: 'string',
    Json: 'Record<string, string|number|boolean|null>',
    Object: 'Record<string, string|number|boolean|null>',
    Hash: 'Record<string, string|number|boolean|null>'
};
export const format = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'string|number|Date',
    Time: 'string|number|Date',
    Datetime: 'string|number|Date',
    Json: 'Record<string, any>',
    Object: 'Record<string, any>',
    Hash: 'Record<string, any>'
};
export const method = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'integer',
    Float: 'float',
    Boolean: 'boolean',
    Date: 'date',
    Datetime: 'date',
    Time: 'date',
    Json: 'object',
    Object: 'object',
    Hash: 'object'
};
export const literal = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'integer',
    Float: 'float',
    Boolean: 'boolean',
    Date: 'date',
    Time: 'time',
    Datetime: 'datetime',
    Json: 'json',
    Object: 'json',
    Hash: 'json'
};
export const mysql = {
    String: 'string',
    Text: 'text',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'date',
    Datetime: 'datetime',
    Time: 'time',
    Json: 'json',
    Object: 'json',
    Hash: 'json'
};
export const pgsql = {
    String: 'string',
    Text: 'text',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'date',
    Datetime: 'timestamp',
    Time: 'time',
    Json: 'jsonb',
    Object: 'jsonb',
    Hash: 'jsonb'
};
export const sqlite = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'real',
    Boolean: 'number',
    Date: 'string',
    Datetime: 'string',
    Time: 'string',
    Json: 'string',
    Object: 'string',
    Hash: 'string'
};
export const helper = {
    String: 'toSqlString',
    Text: 'toSqlString',
    Number: 'toSqlFloat',
    Integer: 'toSqlInteger',
    Float: 'toSqlFloat',
    Boolean: 'toSqlBoolean',
    Date: 'toSqlDate',
    Time: 'toSqlDate',
    Datetime: 'toSqlDate',
    Json: 'toSqlString',
    Object: 'toSqlString',
    Hash: 'toSqlString'
};
