// @stackpress/ingest/cjs/plugin/ImportRouter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class ImportRouter {
    constructor(router, listen) {
        this.imports = new Map();
        this._router = router;
        this._listen = listen;
    }
    action(event, action, priority = 0) {
        var _a;
        if (!this.imports.has(event)) {
            this.imports.set(event, new Set());
        }
        (_a = this.imports.get(event)) === null || _a === void 0 ? void 0 : _a.add({ import: action, priority });
        return function ImportFileAction(req, res, ctx) {
            return __awaiter(this, void 0, void 0, function* () {
                const imports = yield action();
                const callback = imports.default;
                return yield callback(req, res, ctx);
            });
        };
    }
    all(path, action, priority) {
        return this.route('ALL', path, action, priority);
    }
    connect(path, action, priority) {
        return this.route('CONNECT', path, action, priority);
    }
    delete(path, action, priority) {
        return this.route('DELETE', path, action, priority);
    }
    get(path, action, priority) {
        return this.route('GET', path, action, priority);
    }
    head(path, action, priority) {
        return this.route('HEAD', path, action, priority);
    }
    on(event, entry, priority = 0) {
        const key = this._router.eventName(event);
        const action = this.action(key, entry, priority);
        this._listen(key, action, priority);
        return this;
    }
    options(path, action, priority) {
        return this.route('OPTIONS', path, action, priority);
    }
    patch(path, action, priority) {
        return this.route('PATCH', path, action, priority);
    }
    post(path, action, priority) {
        return this.route('POST', path, action, priority);
    }
    put(path, action, priority) {
        return this.route('PUT', path, action, priority);
    }
    route(method, path, entry, priority = 0) {
        const event = this._router.eventName(method, path);
        const action = this.action(event, entry, priority);
        this._listen(event, action, priority);
        return this;
    }
    trace(path, action, priority) {
        return this.route('TRACE', path, action, priority);
    }
    use(router) {
        router.imports.forEach((tasks, event) => {
            if (!this.imports.has(event)) {
                this.imports.set(event, new Set());
            }
            tasks.forEach(task => { var _a; return (_a = this.imports.get(event)) === null || _a === void 0 ? void 0 : _a.add(task); });
        });
        return this;
    }
}
exports.default = ImportRouter;
;
