// frui/cjs/element/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tooltip = exports.Table = exports.Modal = exports.Loader = exports.Badge = exports.Alert = exports.Tgroup = exports.Trow = exports.Tcol = exports.Tfoot = exports.Thead = exports.useModal = exports.ModalProvider = exports.ModalContext = void 0;
var Modal_1 = require("./Modal");
Object.defineProperty(exports, "ModalContext", { enumerable: true, get: function () { return Modal_1.ModalContext; } });
Object.defineProperty(exports, "ModalProvider", { enumerable: true, get: function () { return Modal_1.ModalProvider; } });
Object.defineProperty(exports, "useModal", { enumerable: true, get: function () { return Modal_1.useModal; } });
var Table_1 = require("./Table");
Object.defineProperty(exports, "Thead", { enumerable: true, get: function () { return Table_1.Thead; } });
Object.defineProperty(exports, "Tfoot", { enumerable: true, get: function () { return Table_1.Tfoot; } });
Object.defineProperty(exports, "Tcol", { enumerable: true, get: function () { return Table_1.Tcol; } });
Object.defineProperty(exports, "Trow", { enumerable: true, get: function () { return Table_1.Trow; } });
Object.defineProperty(exports, "Tgroup", { enumerable: true, get: function () { return Table_1.Tgroup; } });
const Alert_1 = __importDefault(require("./Alert"));
exports.Alert = Alert_1.default;
const Badge_1 = __importDefault(require("./Badge"));
exports.Badge = Badge_1.default;
const Loader_1 = __importDefault(require("./Loader"));
exports.Loader = Loader_1.default;
const Modal_2 = __importDefault(require("./Modal"));
exports.Modal = Modal_2.default;
const Table_2 = __importDefault(require("./Table"));
exports.Table = Table_2.default;
const Tooltip_1 = __importDefault(require("./Tooltip"));
exports.Tooltip = Tooltip_1.default;
