// @stackpress/ingest/cjs/whatwg/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhatwgResponse = exports.WhatwgRequest = void 0;
exports.reqToURL = reqToURL;
exports.reqQueryToObject = reqQueryToObject;
exports.readableToReadableStream = readableToReadableStream;
const helpers_js_1 = require("../helpers.js");
exports.WhatwgRequest = global.Request;
exports.WhatwgResponse = global.Response;
function reqToURL(resource) {
    let url = new URL(resource.url);
    url.pathname = url.pathname.replaceAll('//', '/');
    if (url.pathname.endsWith('/')) {
        url.pathname = url.pathname.slice(0, -1);
    }
    return url;
}
;
function reqQueryToObject(resource) {
    return (0, helpers_js_1.objectFromQuery)(reqToURL(resource).searchParams.toString());
}
;
function readableToReadableStream(stream) {
    return new ReadableStream({
        start(controller) {
            stream.on('data', chunk => controller.enqueue(chunk));
            stream.on('end', () => controller.close());
        }
    });
}
;
