// @stackpress/ingest/cjs/http/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dispatcher = exports.loader = exports.Adapter = exports.WriteSession = exports.ReadSession = exports.Server = exports.ViewRouter = exports.ImportRouter = exports.EntryRouter = exports.ActionRouter = exports.Router = exports.Response = exports.Request = exports.PluginLoader = exports.ConfigLoader = exports.Exception = exports.Status = exports.session = exports.cookie = exports.formDataToObject = exports.withUnknownHost = exports.objectFromJson = exports.objectFromFormData = exports.objectFromQuery = exports.isObject = void 0;
exports.gateway = gateway;
exports.handler = handler;
exports.server = server;
exports.router = router;
exports.action = action;
const node_http_1 = require("node:http");
const Router_js_1 = __importDefault(require("../Router.js"));
const Server_js_1 = __importDefault(require("../Server.js"));
const Adapter_js_1 = __importStar(require("./Adapter.js"));
exports.Adapter = Adapter_js_1.default;
Object.defineProperty(exports, "loader", { enumerable: true, get: function () { return Adapter_js_1.loader; } });
Object.defineProperty(exports, "dispatcher", { enumerable: true, get: function () { return Adapter_js_1.dispatcher; } });
var index_js_1 = require("../index.js");
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return index_js_1.isObject; } });
Object.defineProperty(exports, "objectFromQuery", { enumerable: true, get: function () { return index_js_1.objectFromQuery; } });
Object.defineProperty(exports, "objectFromFormData", { enumerable: true, get: function () { return index_js_1.objectFromFormData; } });
Object.defineProperty(exports, "objectFromJson", { enumerable: true, get: function () { return index_js_1.objectFromJson; } });
Object.defineProperty(exports, "withUnknownHost", { enumerable: true, get: function () { return index_js_1.withUnknownHost; } });
Object.defineProperty(exports, "formDataToObject", { enumerable: true, get: function () { return index_js_1.formDataToObject; } });
Object.defineProperty(exports, "cookie", { enumerable: true, get: function () { return index_js_1.cookie; } });
Object.defineProperty(exports, "session", { enumerable: true, get: function () { return index_js_1.session; } });
Object.defineProperty(exports, "Status", { enumerable: true, get: function () { return index_js_1.Status; } });
Object.defineProperty(exports, "Exception", { enumerable: true, get: function () { return index_js_1.Exception; } });
Object.defineProperty(exports, "ConfigLoader", { enumerable: true, get: function () { return index_js_1.ConfigLoader; } });
Object.defineProperty(exports, "PluginLoader", { enumerable: true, get: function () { return index_js_1.PluginLoader; } });
Object.defineProperty(exports, "Request", { enumerable: true, get: function () { return index_js_1.Request; } });
Object.defineProperty(exports, "Response", { enumerable: true, get: function () { return index_js_1.Response; } });
Object.defineProperty(exports, "Router", { enumerable: true, get: function () { return index_js_1.Router; } });
Object.defineProperty(exports, "ActionRouter", { enumerable: true, get: function () { return index_js_1.ActionRouter; } });
Object.defineProperty(exports, "EntryRouter", { enumerable: true, get: function () { return index_js_1.EntryRouter; } });
Object.defineProperty(exports, "ImportRouter", { enumerable: true, get: function () { return index_js_1.ImportRouter; } });
Object.defineProperty(exports, "ViewRouter", { enumerable: true, get: function () { return index_js_1.ViewRouter; } });
Object.defineProperty(exports, "Server", { enumerable: true, get: function () { return index_js_1.Server; } });
Object.defineProperty(exports, "ReadSession", { enumerable: true, get: function () { return index_js_1.ReadSession; } });
Object.defineProperty(exports, "WriteSession", { enumerable: true, get: function () { return index_js_1.WriteSession; } });
function gateway(server) {
    return (options) => (0, node_http_1.createServer)(options, (im, sr) => server.handle(im, sr));
}
;
function handler(context, request, response, action) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield Adapter_js_1.default.plug(context, request, response, action);
    });
}
;
function server(options = {}) {
    options.gateway = options.gateway || gateway;
    options.handler = options.handler || handler;
    return new Server_js_1.default(options);
}
;
function router() {
    return new Router_js_1.default();
}
function action(action) {
    return action;
}
;
action.props = (action) => {
    return function ActionProps(req, res, ctx) {
        return action({ req, res, ctx });
    };
};
