// reactus/esm/server/Document.js
import path from 'node:path';
import { id, renderJSX } from './helpers.js';
export default class Document {
    server;
    entry;
    get id() {
        const hash = id(this.entry, 8);
        const basename = path.basename(this.entry);
        return `${basename}-${hash}`;
    }
    constructor(entry, server) {
        this.entry = entry;
        this.server = server;
    }
    async import() {
        const pagePath = this.server.paths.page;
        const filepath = path.join(pagePath, `${this.id}.js`);
        return await this.server.import(filepath);
    }
    async renderMarkup(props = {}) {
        const { routes, templates } = this.server;
        const document = await this.import();
        const documentTemplate = templates.document;
        const clientRoute = `${routes.client}/${this.id}.js`;
        const cssRoutes = (document.styles || []).map(style => `${routes.css}/${style}`);
        const body = renderJSX(document.default, props);
        const head = renderJSX(document.Head, {
            ...props,
            styles: cssRoutes
        });
        return documentTemplate
            .replace(`<!--document-head-->`, head ?? '')
            .replace(`<!--document-body-->`, body ?? '')
            .replace(`<!--document-props-->`, JSON.stringify(props))
            .replace(`<!--document-client-->`, clientRoute);
    }
}
