// stackpress/cjs/client/Revisions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const Registry_js_1 = __importDefault(require("../schema/Registry.js"));
class Revisions {
    static insert(root, loader, schema) {
        const revisions = new Revisions(root, loader);
        return revisions.insert(schema);
    }
    constructor(root, loader) {
        this._epochs = [];
        this.root = root;
        this.loader = loader;
    }
    first(plus = 0) {
        if (!this._epochs.length) {
            return null;
        }
        return this.index(0 + plus);
    }
    index(index) {
        if (!this._epochs[index]) {
            return null;
        }
        return this.read(this._epochs[index]);
    }
    insert(schema) {
        return __awaiter(this, void 0, void 0, function* () {
            const last = yield this.last();
            const from = last ? JSON.stringify(last.schema, null, 2) : '';
            const to = JSON.stringify(schema, null, 2);
            if (from === to) {
                return this;
            }
            if (!(yield this.loader.fs.exists(this.root))) {
                yield this.loader.fs.mkdir(this.root, { recursive: true });
            }
            const epoch = Date.now();
            this._epochs.push(epoch);
            yield this.loader.fs.writeFile(node_path_1.default.join(this.root, `${epoch}.json`), to);
            return this;
        });
    }
    last() {
        return __awaiter(this, arguments, void 0, function* (minus = 0) {
            if (!this._epochs.length) {
                if (yield this.loader.fs.exists(this.root)) {
                    const glob = yield Promise.resolve().then(() => __importStar(require('fast-glob')));
                    const results = glob.default.sync('*.json', { cwd: this.root });
                    this._epochs = results.map(filename => Number(filename.split('.')[0])).sort();
                }
                else {
                    return null;
                }
            }
            return this.index(this._epochs.length - 1 - Math.abs(minus));
        });
    }
    read(epoch) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._epochs.includes(epoch)) {
                return null;
            }
            const filename = `${epoch}.json`;
            const filepath = node_path_1.default.join(this.root, filename);
            const schema = yield this.loader.import(filepath);
            return {
                date: new Date(epoch),
                file: filename,
                path: filepath,
                schema: schema,
                registry: new Registry_js_1.default(schema)
            };
        });
    }
    size() {
        return this._epochs.length;
    }
}
exports.default = Revisions;
