// @stackpress/inquire/cjs/builder/Insert.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Exception_js_1 = __importDefault(require("../Exception.js"));
class Insert {
    get engine() {
        return this._engine;
    }
    set engine(engine) {
        this._engine = engine;
    }
    constructor(table, engine) {
        this._returning = [];
        this._values = [];
        this._table = table;
        this._engine = engine;
    }
    build() {
        return {
            table: this._table,
            returning: this._returning,
            values: this._values
        };
    }
    query(dialect) {
        var _a;
        dialect = dialect || ((_a = this._engine) === null || _a === void 0 ? void 0 : _a.dialect);
        if (!dialect) {
            throw Exception_js_1.default.for('No dialect provided');
        }
        return dialect.insert(this);
    }
    returning(columns = '*') {
        if (Array.isArray(columns)) {
            this._returning = columns;
        }
        else {
            this._returning = [columns];
        }
        return this;
    }
    then(resolve, reject = () => { }) {
        if (!this._engine) {
            throw Exception_js_1.default.for('No engine provided');
        }
        return this._engine.query(this.query()).then(resolve).catch(reject);
    }
    values(values) {
        if (!Array.isArray(values)) {
            values = [values];
        }
        this._values = values;
        return this;
    }
}
exports.default = Insert;
