// @stackpress/inquire/cjs/builder/Select.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Exception_js_1 = __importDefault(require("../Exception.js"));
class Select {
    get engine() {
        return this._engine;
    }
    set engine(engine) {
        this._engine = engine;
    }
    constructor(select = '*', engine) {
        this._columns = [];
        this._offset = 0;
        this._filters = [];
        this._limit = 0;
        this._relations = [];
        this._sort = [];
        if (Array.isArray(select)) {
            this._columns = select;
        }
        else {
            this._columns = [select];
        }
        this._engine = engine;
    }
    build() {
        return {
            columns: this._columns,
            filters: this._filters,
            limit: this._limit,
            offset: this._offset,
            relations: this._relations,
            sort: this._sort,
            table: this._table
        };
    }
    from(table, as) {
        this._table = [table, as || table];
        return this;
    }
    join(type, table, from, to, as) {
        this._relations.push({ type, table, as: as || table, from, to });
        return this;
    }
    limit(limit) {
        this._limit = limit;
        return this;
    }
    offset(offset) {
        this._offset = offset;
        return this;
    }
    order(column, direction = 'ASC') {
        this._sort.push([column, direction]);
        return this;
    }
    query(dialect) {
        var _a;
        dialect = dialect || ((_a = this._engine) === null || _a === void 0 ? void 0 : _a.dialect);
        if (!dialect) {
            throw Exception_js_1.default.for('No dialect provided');
        }
        return dialect.select(this);
    }
    then(resolve, reject = () => { }) {
        if (!this._engine) {
            throw Exception_js_1.default.for('No engine provided');
        }
        return this._engine.query(this.query()).then(resolve).catch(reject);
    }
    where(query, values = []) {
        this._filters.push([query, values]);
        return this;
    }
}
exports.default = Select;
