// @stackpress/inquire/cjs/builder/Update.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Exception_js_1 = __importDefault(require("../Exception.js"));
class Update {
    get engine() {
        return this._engine;
    }
    set engine(engine) {
        this._engine = engine;
    }
    constructor(table, engine) {
        this._data = {};
        this._filters = [];
        this._table = table;
        this._engine = engine;
    }
    build() {
        return {
            data: this._data,
            filters: this._filters,
            table: this._table
        };
    }
    query(dialect) {
        var _a;
        dialect = dialect || ((_a = this._engine) === null || _a === void 0 ? void 0 : _a.dialect);
        if (!dialect) {
            throw Exception_js_1.default.for('No dialect provided');
        }
        return dialect.update(this);
    }
    set(data) {
        this._data = data;
        return this;
    }
    then(resolve, reject = () => { }) {
        if (!this._engine) {
            throw Exception_js_1.default.for('No engine provided');
        }
        return this._engine.query(this.query()).then(resolve).catch(reject);
    }
    where(query, values = []) {
        this._filters.push([query, values]);
        return this;
    }
}
exports.default = Update;
