// reactus/cjs/server/helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configure = configure;
exports.id = id;
exports.renderJSX = renderJSX;
const node_path_1 = __importDefault(require("node:path"));
const node_crypto_1 = __importDefault(require("node:crypto"));
const react_1 = require("react");
const server_1 = require("react-dom/server");
const jsx_runtime_1 = require("react/jsx-runtime");
const NodeFS_1 = __importDefault(require("@stackpress/lib/NodeFS"));
const constants_js_1 = require("./constants.js");
function configure(options) {
    const cwd = options.cwd || process.cwd();
    return Object.freeze({
        clientRoute: options.clientRoute || '/client',
        cssRoute: options.cssRoute || '/assets',
        cwd: options.cwd || process.cwd(),
        documentTemplate: options.documentTemplate || constants_js_1.DOCUMENT_TEMPLATE,
        fs: options.fs || new NodeFS_1.default(),
        pagePath: options.pagePath || node_path_1.default.join(cwd, '.reactus/page')
    });
}
;
function id(content, length = constants_js_1.HASH_LENGTH) {
    const md5 = node_crypto_1.default.createHash('md5').update(content).digest('hex');
    let num = parseInt(md5.slice(0, 12), 16);
    let hash = '';
    while (num > 0) {
        const index = num % 62;
        hash = constants_js_1.BASE62_ALPHABET[index] + hash;
        num = Math.floor(num / 62);
    }
    return hash.padStart(length, '0').slice(0, length);
}
function renderJSX(element, props = {}) {
    return element ? (0, server_1.renderToString)((0, jsx_runtime_1.jsx)(react_1.StrictMode, { children: (0, jsx_runtime_1.jsx)(element, Object.assign({}, props)) })) : '';
}
