// reactus/cjs/server/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Server = exports.Document = exports.renderJSX = exports.id = exports.configure = exports.DOCUMENT_TEMPLATE = exports.HASH_LENGTH = exports.BASE62_ALPHABET = void 0;
exports.serve = serve;
const Document_js_1 = __importDefault(require("./Document.js"));
exports.Document = Document_js_1.default;
const Server_js_1 = __importDefault(require("./Server.js"));
exports.Server = Server_js_1.default;
const constants_js_1 = require("./constants.js");
Object.defineProperty(exports, "BASE62_ALPHABET", { enumerable: true, get: function () { return constants_js_1.BASE62_ALPHABET; } });
Object.defineProperty(exports, "HASH_LENGTH", { enumerable: true, get: function () { return constants_js_1.HASH_LENGTH; } });
Object.defineProperty(exports, "DOCUMENT_TEMPLATE", { enumerable: true, get: function () { return constants_js_1.DOCUMENT_TEMPLATE; } });
const helpers_js_1 = require("./helpers.js");
Object.defineProperty(exports, "configure", { enumerable: true, get: function () { return helpers_js_1.configure; } });
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return helpers_js_1.id; } });
Object.defineProperty(exports, "renderJSX", { enumerable: true, get: function () { return helpers_js_1.renderJSX; } });
function serve(options) {
    const config = (0, helpers_js_1.configure)(Object.assign({}, options));
    const server = new Server_js_1.default(config);
    return {
        config,
        paths: server.paths,
        routes: server.routes,
        templates: server.templates,
        server,
        render: (entry_1, ...args_1) => __awaiter(this, [entry_1, ...args_1], void 0, function* (entry, props = {}) {
            const document = new Document_js_1.default(entry, server);
            return yield document.renderMarkup(props);
        })
    };
}
