// @stackpress/ingest/esm/plugin/ImportRouter.js
export default class ImportRouter {
    imports = new Map();
    _router;
    _listen;
    constructor(router, listen) {
        this._router = router;
        this._listen = listen;
    }
    action(event, action, priority = 0) {
        if (!this.imports.has(event)) {
            this.imports.set(event, new Set());
        }
        this.imports.get(event)?.add({ import: action, priority });
        return async function ImportFileAction(req, res, ctx) {
            const imports = await action();
            const callback = imports.default;
            return await callback(req, res, ctx);
        };
    }
    all(path, action, priority) {
        return this.route('ALL', path, action, priority);
    }
    connect(path, action, priority) {
        return this.route('CONNECT', path, action, priority);
    }
    delete(path, action, priority) {
        return this.route('DELETE', path, action, priority);
    }
    get(path, action, priority) {
        return this.route('GET', path, action, priority);
    }
    head(path, action, priority) {
        return this.route('HEAD', path, action, priority);
    }
    on(event, entry, priority = 0) {
        const key = this._router.eventName(event);
        const action = this.action(key, entry, priority);
        this._listen(key, action, priority);
        return this;
    }
    options(path, action, priority) {
        return this.route('OPTIONS', path, action, priority);
    }
    patch(path, action, priority) {
        return this.route('PATCH', path, action, priority);
    }
    post(path, action, priority) {
        return this.route('POST', path, action, priority);
    }
    put(path, action, priority) {
        return this.route('PUT', path, action, priority);
    }
    route(method, path, entry, priority = 0) {
        const event = this._router.eventName(method, path);
        const action = this.action(event, entry, priority);
        this._listen(event, action, priority);
        return this;
    }
    trace(path, action, priority) {
        return this.route('TRACE', path, action, priority);
    }
    use(router) {
        router.imports.forEach((tasks, event) => {
            if (!this.imports.has(event)) {
                this.imports.set(event, new Set());
            }
            tasks.forEach(task => this.imports.get(event)?.add(task));
        });
        return this;
    }
}
;
