// @stackpress/lib/cjs/router/Router.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Status_js_1 = __importDefault(require("../Status.js"));
const Nest_js_1 = require("../data/Nest.js");
const ExpressEmitter_js_1 = __importDefault(require("../emitter/ExpressEmitter.js"));
const Request_js_1 = __importDefault(require("./Request.js"));
const Response_js_1 = __importDefault(require("./Response.js"));
class Router extends ExpressEmitter_js_1.default {
    constructor() {
        super('/');
        this.routes = new Map;
    }
    emit(event, req, res) {
        return __awaiter(this, void 0, void 0, function* () {
            const queue = this.tasks(event);
            if (queue.size === 0) {
                return Status_js_1.default.NOT_FOUND;
            }
            return yield queue.run(req, res, this);
        });
    }
    resolve(methodPath, pathRequest, requestResponse, response) {
        return __awaiter(this, void 0, void 0, function* () {
            if (typeof pathRequest === 'string') {
                return this._resolveRoute(methodPath, pathRequest, requestResponse, response);
            }
            return this._resolveEvent(methodPath, pathRequest, requestResponse);
        });
    }
    request(init = {}) {
        return new Request_js_1.default(init);
    }
    response(init = {}) {
        return new Response_js_1.default(init);
    }
    route(method, path, action, priority = 0) {
        const event = this._eventNameFromRoute(method, path);
        this._listen(event, action, priority);
        return this;
    }
    use(emitter) {
        if (emitter instanceof Router) {
            emitter.routes.forEach((route, event) => this.routes.set(event, route));
        }
        super.use(emitter);
        return this;
    }
    _eventNameFromRoute(method, path) {
        method = method.toUpperCase();
        const fragment = this._toFragment(path);
        if (method === 'ANY') {
            const pattern = fragment !== path ? path : '';
            const expression = `^[A-Z]+ ${fragment}${this.separator}*$`;
            const event = this._eventNameFromExpression(expression, pattern);
            this.routes.set(event, { method, path });
            return event;
        }
        let event = `${method} ${path}`;
        if (fragment !== path) {
            const expression = `^${method} ${fragment}${this.separator}*$`;
            event = this._eventNameFromExpression(expression, event);
            this.routes.set(event, { method, path });
            return event;
        }
        this.routes.set(event, { method, path });
        return event;
    }
    _resolveEvent(event, request, response) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!request) {
                request = this.request();
            }
            else if ((0, Nest_js_1.isObject)(request)) {
                const data = request;
                request = this.request({ data });
            }
            const req = request;
            const res = response || this.response();
            yield this.emit(event, req, res);
            return res.toStatusResponse();
        });
    }
    _resolveRoute(method, path, request, response) {
        return __awaiter(this, void 0, void 0, function* () {
            const event = `${method.toUpperCase()} ${path}`;
            return yield this._resolveEvent(event, request, response);
        });
    }
    _task(match, task) {
        return (req, res, ctx) => __awaiter(this, void 0, void 0, function* () {
            this._event = Object.assign(Object.assign(Object.assign({}, match), task), { args: [req, res, ctx], action: task.item });
            req.data.set(match.data.params);
            if (match.data.args.length) {
                req.data.set(match.data.args);
            }
            if (typeof this._before === 'function'
                && (yield this._before(this._event)) === false) {
                return false;
            }
            if ((yield task.item(req, res, ctx)) === false) {
                return false;
            }
            if (typeof this._after === 'function'
                && (yield this._after(this._event)) === false) {
                return false;
            }
        });
    }
}
exports.default = Router;
;
