// frui/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fieldset = exports.Control = exports.Button = exports.Tooltip = exports.Table = exports.Modal = exports.Loader = exports.Badge = exports.Alert = exports.Format = exports.Form = exports.Field = exports.Element = exports.useFieldset = exports.Tgroup = exports.Trow = exports.Tcol = exports.Tfoot = exports.Thead = exports.useModal = exports.ModalProvider = exports.ModalContext = void 0;
const Element = __importStar(require("./element/index.js"));
exports.Element = Element;
const Field = __importStar(require("./field/index.js"));
exports.Field = Field;
const Form = __importStar(require("./form/index.js"));
exports.Form = Form;
const Format = __importStar(require("./format/index.js"));
exports.Format = Format;
var Modal_1 = require("./element/Modal");
Object.defineProperty(exports, "ModalContext", { enumerable: true, get: function () { return Modal_1.ModalContext; } });
Object.defineProperty(exports, "ModalProvider", { enumerable: true, get: function () { return Modal_1.ModalProvider; } });
Object.defineProperty(exports, "useModal", { enumerable: true, get: function () { return Modal_1.useModal; } });
var Table_1 = require("./element/Table");
Object.defineProperty(exports, "Thead", { enumerable: true, get: function () { return Table_1.Thead; } });
Object.defineProperty(exports, "Tfoot", { enumerable: true, get: function () { return Table_1.Tfoot; } });
Object.defineProperty(exports, "Tcol", { enumerable: true, get: function () { return Table_1.Tcol; } });
Object.defineProperty(exports, "Trow", { enumerable: true, get: function () { return Table_1.Trow; } });
Object.defineProperty(exports, "Tgroup", { enumerable: true, get: function () { return Table_1.Tgroup; } });
var Fieldset_1 = require("./form/Fieldset");
Object.defineProperty(exports, "useFieldset", { enumerable: true, get: function () { return Fieldset_1.useFieldset; } });
const Alert_1 = __importDefault(require("./element/Alert"));
exports.Alert = Alert_1.default;
const Badge_1 = __importDefault(require("./element/Badge"));
exports.Badge = Badge_1.default;
const Loader_1 = __importDefault(require("./element/Loader"));
exports.Loader = Loader_1.default;
const Modal_2 = __importDefault(require("./element/Modal"));
exports.Modal = Modal_2.default;
const Table_2 = __importDefault(require("./element/Table"));
exports.Table = Table_2.default;
const Tooltip_1 = __importDefault(require("./element/Tooltip"));
exports.Tooltip = Tooltip_1.default;
const Button_1 = __importDefault(require("./form/Button"));
exports.Button = Button_1.default;
const Control_1 = __importDefault(require("./form/Control"));
exports.Control = Control_1.default;
const Fieldset_2 = __importDefault(require("./form/Fieldset"));
exports.Fieldset = Fieldset_2.default;
