// @stackpress/idea-parser/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginTree = exports.SchemaTree = exports.ModelTree = exports.TypeTree = exports.PropTree = exports.EnumTree = exports.AbstractTree = exports.Compiler = exports.Lexer = exports.Exception = void 0;
exports.final = final;
exports.parse = parse;
const AbstractTree_js_1 = __importDefault(require("./trees/AbstractTree.js"));
exports.AbstractTree = AbstractTree_js_1.default;
const EnumTree_js_1 = __importDefault(require("./trees/EnumTree.js"));
exports.EnumTree = EnumTree_js_1.default;
const PropTree_js_1 = __importDefault(require("./trees/PropTree.js"));
exports.PropTree = PropTree_js_1.default;
const TypeTree_js_1 = __importDefault(require("./trees/TypeTree.js"));
exports.TypeTree = TypeTree_js_1.default;
const ModelTree_js_1 = __importDefault(require("./trees/ModelTree.js"));
exports.ModelTree = ModelTree_js_1.default;
const SchemaTree_js_1 = __importDefault(require("./trees/SchemaTree.js"));
exports.SchemaTree = SchemaTree_js_1.default;
const PluginTree_js_1 = __importDefault(require("./trees/PluginTree.js"));
exports.PluginTree = PluginTree_js_1.default;
const Exception_js_1 = __importDefault(require("./Exception.js"));
exports.Exception = Exception_js_1.default;
const Lexer_js_1 = __importDefault(require("./Lexer.js"));
exports.Lexer = Lexer_js_1.default;
const Compiler_js_1 = __importDefault(require("./Compiler.js"));
exports.Compiler = Compiler_js_1.default;
function final(code) {
    return Compiler_js_1.default.final(SchemaTree_js_1.default.parse(code));
}
function parse(code) {
    return Compiler_js_1.default.schema(SchemaTree_js_1.default.parse(code));
}
