// @stackpress/ingest/cjs/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withUnknownHost = exports.objectFromJson = exports.objectFromFormData = exports.objectFromQuery = exports.isObject = void 0;
exports.formDataToObject = formDataToObject;
const Nest_1 = require("@stackpress/lib/Nest");
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return Nest_1.isObject; } });
Object.defineProperty(exports, "objectFromQuery", { enumerable: true, get: function () { return Nest_1.objectFromQuery; } });
Object.defineProperty(exports, "objectFromFormData", { enumerable: true, get: function () { return Nest_1.objectFromFormData; } });
Object.defineProperty(exports, "objectFromJson", { enumerable: true, get: function () { return Nest_1.objectFromJson; } });
const Request_1 = require("@stackpress/lib/Request");
Object.defineProperty(exports, "withUnknownHost", { enumerable: true, get: function () { return Request_1.withUnknownHost; } });
function formDataToObject(type, body) {
    return type.endsWith('/json')
        ? (0, Nest_1.objectFromJson)(body)
        : type.endsWith('/x-www-form-urlencoded')
            ? (0, Nest_1.objectFromQuery)(body)
            : type.startsWith('multipart/form-data')
                ? (0, Nest_1.objectFromFormData)(body)
                : {};
}
;
