// @stackpress/ingest/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WriteSession = exports.ReadSession = exports.Server = exports.ViewRouter = exports.ImportRouter = exports.EntryRouter = exports.ActionRouter = exports.Router = exports.Response = exports.Request = exports.PluginLoader = exports.ConfigLoader = exports.Exception = exports.Status = exports.server = exports.router = exports.gateway = exports.handler = exports.session = exports.action = exports.cookie = exports.formDataToObject = exports.withUnknownHost = exports.objectFromJson = exports.objectFromFormData = exports.objectFromQuery = exports.isObject = void 0;
const cookie_1 = __importDefault(require("@stackpress/lib/cookie"));
exports.cookie = cookie_1.default;
const Session_1 = require("@stackpress/lib/Session");
Object.defineProperty(exports, "ReadSession", { enumerable: true, get: function () { return Session_1.ReadSession; } });
Object.defineProperty(exports, "WriteSession", { enumerable: true, get: function () { return Session_1.WriteSession; } });
Object.defineProperty(exports, "session", { enumerable: true, get: function () { return Session_1.session; } });
const Status_1 = __importDefault(require("@stackpress/lib/Status"));
exports.Status = Status_1.default;
const Router_js_1 = __importDefault(require("./Router.js"));
exports.Router = Router_js_1.default;
const ActionRouter_js_1 = __importDefault(require("./plugin/ActionRouter.js"));
exports.ActionRouter = ActionRouter_js_1.default;
const EntryRouter_js_1 = __importDefault(require("./plugin/EntryRouter.js"));
exports.EntryRouter = EntryRouter_js_1.default;
const ImportRouter_js_1 = __importDefault(require("./plugin/ImportRouter.js"));
exports.ImportRouter = ImportRouter_js_1.default;
const ViewRouter_js_1 = __importDefault(require("./plugin/ViewRouter.js"));
exports.ViewRouter = ViewRouter_js_1.default;
const Exception_js_1 = __importDefault(require("./Exception.js"));
exports.Exception = Exception_js_1.default;
const Loader_js_1 = require("./Loader.js");
Object.defineProperty(exports, "ConfigLoader", { enumerable: true, get: function () { return Loader_js_1.ConfigLoader; } });
Object.defineProperty(exports, "PluginLoader", { enumerable: true, get: function () { return Loader_js_1.PluginLoader; } });
const Request_js_1 = __importDefault(require("./Request.js"));
exports.Request = Request_js_1.default;
const Response_js_1 = __importDefault(require("./Response.js"));
exports.Response = Response_js_1.default;
const Server_js_1 = __importStar(require("./Server.js"));
exports.Server = Server_js_1.default;
Object.defineProperty(exports, "action", { enumerable: true, get: function () { return Server_js_1.action; } });
Object.defineProperty(exports, "handler", { enumerable: true, get: function () { return Server_js_1.handler; } });
Object.defineProperty(exports, "gateway", { enumerable: true, get: function () { return Server_js_1.gateway; } });
Object.defineProperty(exports, "router", { enumerable: true, get: function () { return Server_js_1.router; } });
Object.defineProperty(exports, "server", { enumerable: true, get: function () { return Server_js_1.server; } });
var helpers_js_1 = require("./helpers.js");
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return helpers_js_1.isObject; } });
Object.defineProperty(exports, "objectFromQuery", { enumerable: true, get: function () { return helpers_js_1.objectFromQuery; } });
Object.defineProperty(exports, "objectFromFormData", { enumerable: true, get: function () { return helpers_js_1.objectFromFormData; } });
Object.defineProperty(exports, "objectFromJson", { enumerable: true, get: function () { return helpers_js_1.objectFromJson; } });
Object.defineProperty(exports, "withUnknownHost", { enumerable: true, get: function () { return helpers_js_1.withUnknownHost; } });
Object.defineProperty(exports, "formDataToObject", { enumerable: true, get: function () { return helpers_js_1.formDataToObject; } });
