// stackpress/cjs/view/server/hooks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRequest = useRequest;
exports.useResponse = useResponse;
exports.useSession = useSession;
exports.useConfig = useConfig;
exports.useServer = useServer;
const react_1 = require("react");
const Nest_1 = require("@stackpress/lib/Nest");
const ServerRequest_js_1 = __importDefault(require("./ServerRequest.js"));
const ServerResponse_js_1 = __importDefault(require("./ServerResponse.js"));
const ServerSession_js_1 = __importDefault(require("./ServerSession.js"));
const ServerContext_js_1 = __importDefault(require("./ServerContext.js"));
function useRequest() {
    const { request } = (0, react_1.useContext)(ServerContext_js_1.default);
    return new ServerRequest_js_1.default(request);
}
function useResponse() {
    const { response } = (0, react_1.useContext)(ServerContext_js_1.default);
    return new ServerResponse_js_1.default(response);
}
function useSession() {
    const { session } = (0, react_1.useContext)(ServerContext_js_1.default);
    return new ServerSession_js_1.default(session);
}
function useConfig() {
    const { data } = (0, react_1.useContext)(ServerContext_js_1.default);
    return (0, Nest_1.nest)(data);
}
function useServer() {
    const { data, request, response, session } = (0, react_1.useContext)(ServerContext_js_1.default);
    return {
        config: (0, Nest_1.nest)(data),
        request: new ServerRequest_js_1.default(request),
        response: new ServerResponse_js_1.default(response),
        session: new ServerSession_js_1.default(session)
    };
}
