// stackpress/cjs/view/server/ServerContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = exports.unknownHost = void 0;
const react_1 = require("react");
const Request_1 = require("@stackpress/lib/Request");
exports.unknownHost = new URL((0, Request_1.withUnknownHost)('/'));
exports.config = {
    data: {},
    session: {
        id: '',
        name: 'Guest',
        roles: ['GUEST'],
        token: '',
        permits: []
    },
    request: {
        url: {
            hash: exports.unknownHost.hash,
            host: exports.unknownHost.host,
            hostname: exports.unknownHost.hostname,
            href: exports.unknownHost.href,
            origin: exports.unknownHost.origin,
            pathname: exports.unknownHost.pathname,
            port: exports.unknownHost.port,
            protocol: exports.unknownHost.protocol,
            search: exports.unknownHost.search
        },
        headers: {},
        session: {},
        method: 'GET',
        mime: '',
        data: {}
    },
    response: {
        code: 0,
        status: ''
    }
};
const ServerContext = (0, react_1.createContext)(exports.config);
exports.default = ServerContext;
