// stackpress/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadonlyPath = exports.ReadonlyNest = exports.ReadonlyMap = exports.QueryString = exports.PathString = exports.FormData = exports.FileData = exports.ArgString = exports.render = exports.snakerize = exports.lowerize = exports.hash = exports.encrypt = exports.decrypt = exports.dasherize = exports.capitalize = exports.camelize = exports.jsonCompare = exports.Server = exports.Document = exports.Builder = exports.VirtualServer = exports.ServerResource = exports.ServerManifest = exports.ServerLoader = exports.DocumentRender = exports.DocumentLoader = exports.DocumentBuilder = exports.viteVFSPlugin = exports.viteHMRPlugin = exports.viteFilePlugin = exports.viteCSSPlugin = exports.renderJSX = exports.fileHash = exports.CLIENT_TEMPLATE = exports.PAGE_TEMPLATE = exports.DOCUMENT_TEMPLATE = exports.HASH_LENGTH = exports.BASE62_ALPHABET = exports.VFS_RESOLVED = exports.VFS_PROTOCOL = exports.Registry = exports.assert = exports.Model = exports.Fieldset = exports.Column = exports.Attributes = exports.objectToAttributeString = exports.generators = exports.Revisions = void 0;
exports.terminal = exports.scripts = exports.sql = exports.Exception = exports.I18N = exports.Session = exports.readableToReadableStream = exports.reqQueryToObject = exports.reqToURL = exports.readableStreamToReadable = exports.imQueryToObject = exports.imToURL = exports.formDataToObject = exports.shouldBeAnArray = exports.makeObject = exports.makeArray = exports.session = exports.control = exports.nest = exports.set = exports.map = exports.codes = exports.Status = exports.Reflection = exports.Transformer = exports.Terminal = exports.RouteEmitter = exports.ExpressEmitter = exports.EventEmitter = exports.TaskQueue = exports.ItemQueue = exports.NodeFS = exports.FileLoader = exports.Nest = exports.ReadonlySet = void 0;
const Session_js_1 = __importDefault(require("./session/Session.js"));
exports.Session = Session_js_1.default;
const Language_js_1 = __importDefault(require("./language/Language.js"));
exports.I18N = Language_js_1.default;
const Exception_js_1 = __importDefault(require("./Exception.js"));
exports.Exception = Exception_js_1.default;
var index_js_1 = require("./client/index.js");
Object.defineProperty(exports, "Revisions", { enumerable: true, get: function () { return index_js_1.Revisions; } });
var index_js_2 = require("./schema/index.js");
Object.defineProperty(exports, "generators", { enumerable: true, get: function () { return index_js_2.generators; } });
Object.defineProperty(exports, "objectToAttributeString", { enumerable: true, get: function () { return index_js_2.objectToAttributeString; } });
Object.defineProperty(exports, "Attributes", { enumerable: true, get: function () { return index_js_2.Attributes; } });
Object.defineProperty(exports, "Column", { enumerable: true, get: function () { return index_js_2.Column; } });
Object.defineProperty(exports, "Fieldset", { enumerable: true, get: function () { return index_js_2.Fieldset; } });
Object.defineProperty(exports, "Model", { enumerable: true, get: function () { return index_js_2.Model; } });
Object.defineProperty(exports, "assert", { enumerable: true, get: function () { return index_js_2.assert; } });
Object.defineProperty(exports, "Registry", { enumerable: true, get: function () { return index_js_2.Registry; } });
var reactus_1 = require("reactus");
Object.defineProperty(exports, "VFS_PROTOCOL", { enumerable: true, get: function () { return reactus_1.VFS_PROTOCOL; } });
Object.defineProperty(exports, "VFS_RESOLVED", { enumerable: true, get: function () { return reactus_1.VFS_RESOLVED; } });
Object.defineProperty(exports, "BASE62_ALPHABET", { enumerable: true, get: function () { return reactus_1.BASE62_ALPHABET; } });
Object.defineProperty(exports, "HASH_LENGTH", { enumerable: true, get: function () { return reactus_1.HASH_LENGTH; } });
Object.defineProperty(exports, "DOCUMENT_TEMPLATE", { enumerable: true, get: function () { return reactus_1.DOCUMENT_TEMPLATE; } });
Object.defineProperty(exports, "PAGE_TEMPLATE", { enumerable: true, get: function () { return reactus_1.PAGE_TEMPLATE; } });
Object.defineProperty(exports, "CLIENT_TEMPLATE", { enumerable: true, get: function () { return reactus_1.CLIENT_TEMPLATE; } });
Object.defineProperty(exports, "fileHash", { enumerable: true, get: function () { return reactus_1.id; } });
Object.defineProperty(exports, "renderJSX", { enumerable: true, get: function () { return reactus_1.renderJSX; } });
Object.defineProperty(exports, "viteCSSPlugin", { enumerable: true, get: function () { return reactus_1.css; } });
Object.defineProperty(exports, "viteFilePlugin", { enumerable: true, get: function () { return reactus_1.file; } });
Object.defineProperty(exports, "viteHMRPlugin", { enumerable: true, get: function () { return reactus_1.hmr; } });
Object.defineProperty(exports, "viteVFSPlugin", { enumerable: true, get: function () { return reactus_1.vfs; } });
Object.defineProperty(exports, "DocumentBuilder", { enumerable: true, get: function () { return reactus_1.DocumentBuilder; } });
Object.defineProperty(exports, "DocumentLoader", { enumerable: true, get: function () { return reactus_1.DocumentLoader; } });
Object.defineProperty(exports, "DocumentRender", { enumerable: true, get: function () { return reactus_1.DocumentRender; } });
Object.defineProperty(exports, "ServerLoader", { enumerable: true, get: function () { return reactus_1.ServerLoader; } });
Object.defineProperty(exports, "ServerManifest", { enumerable: true, get: function () { return reactus_1.ServerManifest; } });
Object.defineProperty(exports, "ServerResource", { enumerable: true, get: function () { return reactus_1.ServerResource; } });
Object.defineProperty(exports, "VirtualServer", { enumerable: true, get: function () { return reactus_1.VirtualServer; } });
Object.defineProperty(exports, "Builder", { enumerable: true, get: function () { return reactus_1.Builder; } });
Object.defineProperty(exports, "Document", { enumerable: true, get: function () { return reactus_1.Document; } });
Object.defineProperty(exports, "Server", { enumerable: true, get: function () { return reactus_1.Server; } });
var helpers_1 = require("@stackpress/inquire/helpers");
Object.defineProperty(exports, "jsonCompare", { enumerable: true, get: function () { return helpers_1.jsonCompare; } });
var helpers_js_1 = require("./schema/helpers.js");
Object.defineProperty(exports, "camelize", { enumerable: true, get: function () { return helpers_js_1.camelize; } });
Object.defineProperty(exports, "capitalize", { enumerable: true, get: function () { return helpers_js_1.capitalize; } });
Object.defineProperty(exports, "dasherize", { enumerable: true, get: function () { return helpers_js_1.dasherize; } });
Object.defineProperty(exports, "decrypt", { enumerable: true, get: function () { return helpers_js_1.decrypt; } });
Object.defineProperty(exports, "encrypt", { enumerable: true, get: function () { return helpers_js_1.encrypt; } });
Object.defineProperty(exports, "hash", { enumerable: true, get: function () { return helpers_js_1.hash; } });
Object.defineProperty(exports, "lowerize", { enumerable: true, get: function () { return helpers_js_1.lowerize; } });
Object.defineProperty(exports, "snakerize", { enumerable: true, get: function () { return helpers_js_1.snakerize; } });
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return helpers_js_1.render; } });
var lib_js_1 = require("./lib.js");
Object.defineProperty(exports, "ArgString", { enumerable: true, get: function () { return lib_js_1.ArgString; } });
Object.defineProperty(exports, "FileData", { enumerable: true, get: function () { return lib_js_1.FileData; } });
Object.defineProperty(exports, "FormData", { enumerable: true, get: function () { return lib_js_1.FormData; } });
Object.defineProperty(exports, "PathString", { enumerable: true, get: function () { return lib_js_1.PathString; } });
Object.defineProperty(exports, "QueryString", { enumerable: true, get: function () { return lib_js_1.QueryString; } });
Object.defineProperty(exports, "ReadonlyMap", { enumerable: true, get: function () { return lib_js_1.ReadonlyMap; } });
Object.defineProperty(exports, "ReadonlyNest", { enumerable: true, get: function () { return lib_js_1.ReadonlyNest; } });
Object.defineProperty(exports, "ReadonlyPath", { enumerable: true, get: function () { return lib_js_1.ReadonlyPath; } });
Object.defineProperty(exports, "ReadonlySet", { enumerable: true, get: function () { return lib_js_1.ReadonlySet; } });
Object.defineProperty(exports, "Nest", { enumerable: true, get: function () { return lib_js_1.Nest; } });
Object.defineProperty(exports, "FileLoader", { enumerable: true, get: function () { return lib_js_1.FileLoader; } });
Object.defineProperty(exports, "NodeFS", { enumerable: true, get: function () { return lib_js_1.NodeFS; } });
Object.defineProperty(exports, "ItemQueue", { enumerable: true, get: function () { return lib_js_1.ItemQueue; } });
Object.defineProperty(exports, "TaskQueue", { enumerable: true, get: function () { return lib_js_1.TaskQueue; } });
Object.defineProperty(exports, "EventEmitter", { enumerable: true, get: function () { return lib_js_1.EventEmitter; } });
Object.defineProperty(exports, "ExpressEmitter", { enumerable: true, get: function () { return lib_js_1.ExpressEmitter; } });
Object.defineProperty(exports, "RouteEmitter", { enumerable: true, get: function () { return lib_js_1.RouteEmitter; } });
Object.defineProperty(exports, "Terminal", { enumerable: true, get: function () { return lib_js_1.Terminal; } });
Object.defineProperty(exports, "Transformer", { enumerable: true, get: function () { return lib_js_1.Transformer; } });
Object.defineProperty(exports, "Reflection", { enumerable: true, get: function () { return lib_js_1.Reflection; } });
Object.defineProperty(exports, "Status", { enumerable: true, get: function () { return lib_js_1.Status; } });
Object.defineProperty(exports, "codes", { enumerable: true, get: function () { return lib_js_1.codes; } });
Object.defineProperty(exports, "map", { enumerable: true, get: function () { return lib_js_1.map; } });
Object.defineProperty(exports, "set", { enumerable: true, get: function () { return lib_js_1.set; } });
Object.defineProperty(exports, "nest", { enumerable: true, get: function () { return lib_js_1.nest; } });
Object.defineProperty(exports, "control", { enumerable: true, get: function () { return lib_js_1.control; } });
Object.defineProperty(exports, "session", { enumerable: true, get: function () { return lib_js_1.session; } });
Object.defineProperty(exports, "makeArray", { enumerable: true, get: function () { return lib_js_1.makeArray; } });
Object.defineProperty(exports, "makeObject", { enumerable: true, get: function () { return lib_js_1.makeObject; } });
Object.defineProperty(exports, "shouldBeAnArray", { enumerable: true, get: function () { return lib_js_1.shouldBeAnArray; } });
var helpers_2 = require("@stackpress/ingest/helpers");
Object.defineProperty(exports, "formDataToObject", { enumerable: true, get: function () { return helpers_2.formDataToObject; } });
var helpers_3 = require("@stackpress/ingest/http/helpers");
Object.defineProperty(exports, "imToURL", { enumerable: true, get: function () { return helpers_3.imToURL; } });
Object.defineProperty(exports, "imQueryToObject", { enumerable: true, get: function () { return helpers_3.imQueryToObject; } });
Object.defineProperty(exports, "readableStreamToReadable", { enumerable: true, get: function () { return helpers_3.readableStreamToReadable; } });
var helpers_4 = require("@stackpress/ingest/whatwg/helpers");
Object.defineProperty(exports, "reqToURL", { enumerable: true, get: function () { return helpers_4.reqToURL; } });
Object.defineProperty(exports, "reqQueryToObject", { enumerable: true, get: function () { return helpers_4.reqQueryToObject; } });
Object.defineProperty(exports, "readableToReadableStream", { enumerable: true, get: function () { return helpers_4.readableToReadableStream; } });
const sql = __importStar(require("./sql/index.js"));
exports.sql = sql;
const scripts = __importStar(require("./scripts/index.js"));
exports.scripts = scripts;
const terminal = __importStar(require("./terminal/index.js"));
exports.terminal = terminal;
