// reactus/cjs/Document.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const DocumentBuilder_js_1 = __importDefault(require("./DocumentBuilder.js"));
const DocumentLoader_js_1 = __importDefault(require("./DocumentLoader.js"));
const DocumentRender_js_1 = __importDefault(require("./DocumentRender.js"));
const helpers_js_1 = require("./helpers.js");
class Document {
    get id() {
        const hash = (0, helpers_js_1.id)(this.entry, 8);
        const basename = node_path_1.default.basename(this.entry);
        return `${basename}-${hash}`;
    }
    constructor(entry, server) {
        this.entry = entry;
        this.server = server;
        this.builder = new DocumentBuilder_js_1.default(this);
        this.loader = new DocumentLoader_js_1.default(this);
        this.render = new DocumentRender_js_1.default(this);
    }
}
exports.default = Document;
