// stackpress/cjs/view/config/production.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = config;
const server_1 = require("reactus/server");
const Status_1 = __importDefault(require("@stackpress/lib/Status"));
function config(server) {
    const cwd = server.config.path('server.cwd', process.cwd());
    const options = server.config.path('view.engine', {});
    const engine = (0, server_1.serve)((0, server_1.configure)(Object.assign(Object.assign({}, options), { cwd })));
    server.register('reactus', engine);
    server.view.render = (action, props) => engine.render(action, props);
    server.view.engine = (action, req, res, ctx) => __awaiter(this, void 0, void 0, function* () {
        const status = Status_1.default.get(res.code || 200);
        res.setStatus(status.code, status.status);
        const noview = ctx.config.path('view.noview', 'json');
        if (res.redirected
            || req.data.has(noview)
            || typeof res.body === 'string')
            return;
        const props = ctx.config.path('view.props', {});
        const session = yield ctx.resolve('me', req);
        const html = yield ctx.view.render(action, {
            data: Object.assign(Object.assign({}, props), res.data()),
            session: session.results,
            request: {
                url: {
                    hash: req.url.hash,
                    host: req.url.host,
                    hostname: req.url.hostname,
                    href: req.url.href,
                    origin: req.url.origin,
                    pathname: req.url.pathname,
                    port: req.url.port,
                    protocol: req.url.protocol,
                    search: req.url.search
                },
                headers: Object.fromEntries(req.headers.entries()),
                session: req.session.data,
                method: req.method,
                mime: req.mimetype,
                data: req.data()
            },
            response: res.toStatusResponse()
        });
        if (html) {
            res.setHTML(html, status.code, status.status);
        }
    });
}
;
