// stackpress/cjs/view/client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pagination = exports.Crumbs = exports.ThemeProvider = exports.ThemeContext = exports.ModalProvider = exports.ModalContext = exports.ModalConfirm = exports.LayoutRight = exports.LayoutProvider = exports.LayoutAdmin = exports.LayoutPanel = exports.AdminUserMenu = exports.AdminApp = exports.BlankApp = exports.PanelApp = exports.LayoutMain = exports.LayoutLeft = exports.LayoutHead = exports.LayoutBlank = exports.ServerSession = exports.ServerResponse = exports.ServerRequest = exports.ServerProvider = exports.ServerContext = exports.StackpressProvider = exports.matchRoute = exports.matchEvent = exports.matchAnyRoute = exports.matchAnyEvent = exports.useServer = exports.useSession = exports.useConfig = exports.useResponse = exports.useRequest = exports.useTheme = exports.unload = exports.flash = exports.notify = exports.useNotify = exports.useToggle = exports.useConfirm = exports.useModal = exports.filter = exports.order = exports.paginate = exports.useStripe = exports.useLanguage = exports.Translate = exports.R22nProvider = exports.R22nContext = void 0;
exports.unknownHost = exports.defaultNotifyConfig = exports.defaultServerConfig = exports.NotifyProvider = exports.NotifyContext = exports.NotifyContainer = void 0;
const ServerContext_js_1 = __importStar(require("./server/ServerContext.js"));
exports.ServerContext = ServerContext_js_1.default;
Object.defineProperty(exports, "unknownHost", { enumerable: true, get: function () { return ServerContext_js_1.unknownHost; } });
Object.defineProperty(exports, "defaultServerConfig", { enumerable: true, get: function () { return ServerContext_js_1.config; } });
const ServerProvider_js_1 = __importDefault(require("./server/ServerProvider.js"));
exports.ServerProvider = ServerProvider_js_1.default;
const ServerRequest_js_1 = __importDefault(require("./server/ServerRequest.js"));
exports.ServerRequest = ServerRequest_js_1.default;
const ServerResponse_js_1 = __importDefault(require("./server/ServerResponse.js"));
exports.ServerResponse = ServerResponse_js_1.default;
const ServerSession_js_1 = __importDefault(require("./server/ServerSession.js"));
exports.ServerSession = ServerSession_js_1.default;
const Crumbs_js_1 = __importDefault(require("./element/Crumbs.js"));
exports.Crumbs = Crumbs_js_1.default;
const Pagination_js_1 = __importDefault(require("./element/Pagination.js"));
exports.Pagination = Pagination_js_1.default;
const LayoutHead_js_1 = __importDefault(require("./layout/components/LayoutHead.js"));
exports.LayoutHead = LayoutHead_js_1.default;
const LayoutLeft_js_1 = __importDefault(require("./layout/components/LayoutLeft.js"));
exports.LayoutLeft = LayoutLeft_js_1.default;
const LayoutMain_js_1 = __importDefault(require("./layout/components/LayoutMain.js"));
exports.LayoutMain = LayoutMain_js_1.default;
const LayoutRight_js_1 = __importDefault(require("./layout/components/LayoutRight.js"));
exports.LayoutRight = LayoutRight_js_1.default;
const LayoutBlank_js_1 = __importStar(require("./layout/LayoutBlank.js"));
exports.LayoutBlank = LayoutBlank_js_1.default;
Object.defineProperty(exports, "BlankApp", { enumerable: true, get: function () { return LayoutBlank_js_1.BlankApp; } });
const LayoutPanel_js_1 = __importStar(require("./layout/LayoutPanel.js"));
exports.LayoutPanel = LayoutPanel_js_1.default;
Object.defineProperty(exports, "PanelApp", { enumerable: true, get: function () { return LayoutPanel_js_1.PanelApp; } });
const LayoutAdmin_js_1 = __importStar(require("./layout/LayoutAdmin.js"));
exports.LayoutAdmin = LayoutAdmin_js_1.default;
Object.defineProperty(exports, "AdminApp", { enumerable: true, get: function () { return LayoutAdmin_js_1.AdminApp; } });
Object.defineProperty(exports, "AdminUserMenu", { enumerable: true, get: function () { return LayoutAdmin_js_1.AdminUserMenu; } });
const LayoutProvider_js_1 = __importDefault(require("./layout/LayoutProvider.js"));
exports.LayoutProvider = LayoutProvider_js_1.default;
const ModalConfirm_js_1 = __importDefault(require("./modal/ModalConfirm.js"));
exports.ModalConfirm = ModalConfirm_js_1.default;
const ModalContext_js_1 = __importDefault(require("./modal/ModalContext.js"));
exports.ModalContext = ModalContext_js_1.default;
const ModalProvider_js_1 = __importDefault(require("./modal/ModalProvider.js"));
exports.ModalProvider = ModalProvider_js_1.default;
const NotifyContainer_js_1 = __importDefault(require("./notify/NotifyContainer.js"));
exports.NotifyContainer = NotifyContainer_js_1.default;
const NotifyContext_js_1 = __importStar(require("./notify/NotifyContext.js"));
exports.NotifyContext = NotifyContext_js_1.default;
Object.defineProperty(exports, "defaultNotifyConfig", { enumerable: true, get: function () { return NotifyContext_js_1.config; } });
const NotifyProvider_js_1 = __importDefault(require("./notify/NotifyProvider.js"));
exports.NotifyProvider = NotifyProvider_js_1.default;
const ThemeContext_js_1 = __importDefault(require("./theme/ThemeContext.js"));
exports.ThemeContext = ThemeContext_js_1.default;
const ThemeProvider_js_1 = __importDefault(require("./theme/ThemeProvider.js"));
exports.ThemeProvider = ThemeProvider_js_1.default;
var r22n_1 = require("r22n");
Object.defineProperty(exports, "R22nContext", { enumerable: true, get: function () { return r22n_1.R22nContext; } });
Object.defineProperty(exports, "R22nProvider", { enumerable: true, get: function () { return r22n_1.R22nProvider; } });
Object.defineProperty(exports, "Translate", { enumerable: true, get: function () { return r22n_1.Translate; } });
Object.defineProperty(exports, "useLanguage", { enumerable: true, get: function () { return r22n_1.useLanguage; } });
var hooks_js_1 = require("./hooks.js");
Object.defineProperty(exports, "useStripe", { enumerable: true, get: function () { return hooks_js_1.useStripe; } });
var helpers_js_1 = require("./helpers.js");
Object.defineProperty(exports, "paginate", { enumerable: true, get: function () { return helpers_js_1.paginate; } });
Object.defineProperty(exports, "order", { enumerable: true, get: function () { return helpers_js_1.order; } });
Object.defineProperty(exports, "filter", { enumerable: true, get: function () { return helpers_js_1.filter; } });
var hooks_js_2 = require("./modal/hooks.js");
Object.defineProperty(exports, "useModal", { enumerable: true, get: function () { return hooks_js_2.useModal; } });
Object.defineProperty(exports, "useConfirm", { enumerable: true, get: function () { return hooks_js_2.useConfirm; } });
var hooks_js_3 = require("./layout/hooks.js");
Object.defineProperty(exports, "useToggle", { enumerable: true, get: function () { return hooks_js_3.useToggle; } });
var hooks_js_4 = require("./notify/hooks.js");
Object.defineProperty(exports, "useNotify", { enumerable: true, get: function () { return hooks_js_4.useNotify; } });
Object.defineProperty(exports, "notify", { enumerable: true, get: function () { return hooks_js_4.notify; } });
Object.defineProperty(exports, "flash", { enumerable: true, get: function () { return hooks_js_4.flash; } });
Object.defineProperty(exports, "unload", { enumerable: true, get: function () { return hooks_js_4.unload; } });
var hooks_js_5 = require("./theme/hooks.js");
Object.defineProperty(exports, "useTheme", { enumerable: true, get: function () { return hooks_js_5.useTheme; } });
var hooks_js_6 = require("./server/hooks.js");
Object.defineProperty(exports, "useRequest", { enumerable: true, get: function () { return hooks_js_6.useRequest; } });
Object.defineProperty(exports, "useResponse", { enumerable: true, get: function () { return hooks_js_6.useResponse; } });
Object.defineProperty(exports, "useConfig", { enumerable: true, get: function () { return hooks_js_6.useConfig; } });
Object.defineProperty(exports, "useSession", { enumerable: true, get: function () { return hooks_js_6.useSession; } });
Object.defineProperty(exports, "useServer", { enumerable: true, get: function () { return hooks_js_6.useServer; } });
var helpers_js_2 = require("./server/helpers.js");
Object.defineProperty(exports, "matchAnyEvent", { enumerable: true, get: function () { return helpers_js_2.matchAnyEvent; } });
Object.defineProperty(exports, "matchAnyRoute", { enumerable: true, get: function () { return helpers_js_2.matchAnyRoute; } });
Object.defineProperty(exports, "matchEvent", { enumerable: true, get: function () { return helpers_js_2.matchEvent; } });
Object.defineProperty(exports, "matchRoute", { enumerable: true, get: function () { return helpers_js_2.matchRoute; } });
exports.StackpressProvider = LayoutProvider_js_1.default;
