// @stackpress/inquire/esm/builder/Create.js
import Exception from '../Exception.js';
export default class Create {
    _engine;
    _fields = {};
    _foreign = {};
    _keys = {};
    _primary = [];
    _table;
    _unique = {};
    get engine() {
        return this._engine;
    }
    set engine(engine) {
        this._engine = engine;
    }
    constructor(table, engine) {
        this._table = table;
        this._engine = engine;
    }
    addField(name, field) {
        this._fields[name] = field;
        return this;
    }
    addForeignKey(name, foriegnKey) {
        this._foreign[name] = foriegnKey;
        return this;
    }
    addKey(name, field) {
        if (!Array.isArray(field)) {
            field = [field];
        }
        this._keys[name] = field;
        return this;
    }
    addPrimaryKey(name) {
        this._primary.push(name);
        return this;
    }
    addUniqueKey(name, field) {
        if (!Array.isArray(field)) {
            field = [field];
        }
        this._unique[name] = field;
        return this;
    }
    build() {
        return {
            fields: this._fields,
            foreign: this._foreign,
            keys: this._keys,
            primary: this._primary,
            table: this._table,
            unique: this._unique
        };
    }
    query(dialect) {
        dialect = dialect || this._engine?.dialect;
        if (!dialect) {
            throw Exception.for('No dialect provided');
        }
        return dialect.create(this);
    }
    then(resolve, reject = () => { }) {
        if (!this._engine) {
            throw Exception.for('No engine provided');
        }
        const queries = this.query();
        const last = queries.pop();
        return this._engine.transaction(async (connection) => {
            for (const request of queries) {
                await connection.query(connection.format(request));
            }
            return await connection.query(connection.format(last));
        }).then(resolve).catch(reject);
    }
}
