// @stackpress/inquire/esm/builder/Insert.js
import Exception from '../Exception.js';
export default class Insert {
    _engine;
    _returning = [];
    _table;
    _values = [];
    get engine() {
        return this._engine;
    }
    set engine(engine) {
        this._engine = engine;
    }
    constructor(table, engine) {
        this._table = table;
        this._engine = engine;
    }
    build() {
        return {
            table: this._table,
            returning: this._returning,
            values: this._values
        };
    }
    query(dialect) {
        dialect = dialect || this._engine?.dialect;
        if (!dialect) {
            throw Exception.for('No dialect provided');
        }
        return dialect.insert(this);
    }
    returning(columns = '*') {
        if (Array.isArray(columns)) {
            this._returning = columns;
        }
        else {
            this._returning = [columns];
        }
        return this;
    }
    then(resolve, reject = () => { }) {
        if (!this._engine) {
            throw Exception.for('No engine provided');
        }
        return this._engine.query(this.query()).then(resolve).catch(reject);
    }
    values(values) {
        if (!Array.isArray(values)) {
            values = [values];
        }
        this._values = values;
        return this;
    }
}
