// frui/esm/field/Switch.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, useEffect } from 'react';
export function useSwitch(config) {
    const { onChange, onUpdate, defaultChecked, checked } = config;
    const [isChecked, check] = useState(Boolean(defaultChecked || checked));
    const [isHovering, hover] = useState(false);
    useEffect(() => {
        if (typeof checked === 'undefined')
            return;
        if (checked !== isChecked) {
            check(checked);
        }
    }, [checked]);
    return {
        isHovering,
        isChecked,
        handlers: {
            out: (_e) => hover(false),
            over: (_e) => hover(true),
            change: (e) => {
                if (e.target.checked !== checked) {
                    check(e.target.checked);
                }
                onChange && onChange(e);
                onUpdate && onUpdate(e.target.value, e.target.checked);
            }
        }
    };
}
export default function Switch(props) {
    const { defaultChecked, checked, label, error, rounded, onoff, yesno, checkex, sunmoon, ridge, blue, orange, green, style, className, onChange, onUpdate, ...attributes } = props;
    const { handlers } = useSwitch({
        onChange,
        onUpdate,
        checked,
        defaultChecked
    });
    const classNames = ['frui-field-switch'];
    if (rounded) {
        classNames.push('frui-field-switch-rounded');
    }
    if (error) {
        classNames.push('frui-tx-error');
    }
    if (onoff) {
        classNames.push('frui-field-switch-onoff');
    }
    else if (yesno) {
        classNames.push('frui-field-switch-yesno');
    }
    else if (sunmoon) {
        classNames.push('frui-field-switch-sunmoon');
    }
    else {
        classNames.push('frui-field-switch-checkex');
    }
    if (ridge) {
        classNames.push('frui-field-switch-ridge');
    }
    else {
        classNames.push('frui-field-switch-smooth');
    }
    if (blue) {
        classNames.push('frui-field-switch-blue');
    }
    else if (orange) {
        classNames.push('frui-field-switch-orange');
    }
    else if (green) {
        classNames.push('frui-field-switch-green');
    }
    else {
        classNames.push('frui-field-switch-default');
    }
    if (className) {
        classNames.push(className);
    }
    return (_jsxs("label", { className: classNames.join(' '), style: style, children: [_jsx("input", { ...attributes, onChange: handlers.change, onMouseOut: handlers.out, onMouseOver: handlers.over, type: "checkbox", className: "frui-field-switch-control", checked: checked, defaultChecked: defaultChecked }), _jsx("span", { className: "frui-field-switch-label", children: label })] }));
}
