// frui/esm/field/Taglist.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, useEffect } from 'react';
export function useTaglist(config) {
    const { value, defaultValue, onChange, onUpdate } = config;
    const [input, setInput] = useState('');
    const [isKeyReleased, setIsKeyReleased] = useState(false);
    const [tags, setTags] = useState(defaultValue || []);
    useEffect(() => {
        if (Array.isArray(value)
            && JSON.stringify(value) === JSON.stringify(tags))
            return;
        onUpdate && onUpdate(tags);
    }, [tags]);
    useEffect(() => {
        if (!Array.isArray(value))
            return;
        setTags(value);
    }, [value]);
    return {
        input,
        tags,
        handlers: {
            change: (e) => {
                const { value } = e.target;
                setInput(value);
                onChange && onChange(e);
            },
            remove: (index) => {
                setTags(prevState => prevState.filter((_tag, i) => i !== index));
            },
            edit: (e) => {
                const { key } = e;
                const trimmedInput = input.trim();
                if ((key === 'Enter' || key === ',') && trimmedInput.length && !tags.includes(trimmedInput)) {
                    e.preventDefault();
                    setTags(prevState => [...prevState, trimmedInput]);
                    setInput('');
                }
                if (key === 'Backspace' && !input.length && tags.length && isKeyReleased) {
                    const tagsCopy = [...tags];
                    const poppedTag = tagsCopy.pop();
                    e.preventDefault();
                    setTags(tagsCopy);
                    setInput(poppedTag || '');
                }
                setIsKeyReleased(false);
            },
            save: () => {
                setIsKeyReleased(true);
            }
        }
    };
}
;
export default function Taglist(props) {
    const { name, value, defaultValue, color, info, warning, success, danger, muted, error, className, style, onChange, onUpdate, ...attributes } = props;
    const { input, tags, handlers } = useTaglist({
        onChange,
        onUpdate,
        value: Array.isArray(value)
            ? value
            : typeof value === 'string'
                ? value.split(',')
                : undefined,
        defaultValue: Array.isArray(defaultValue)
            ? defaultValue
            : typeof defaultValue === 'string'
                ? defaultValue.split(',')
                : undefined,
    });
    const classNames = ['frui-field-taglist'];
    if (error) {
        classNames.push('frui-tx-error');
    }
    if (className) {
        classNames.push(className);
    }
    const tagClasses = ['frui-field-taglist-tag'];
    const tagStyle = {};
    if (color) {
        tagStyle.backgroundColor = color;
    }
    else if (info) {
        tagClasses.push('frui-bg-info');
    }
    else if (warning) {
        tagClasses.push('frui-bg-warning');
    }
    else if (success) {
        tagClasses.push('frui-bg-success');
    }
    else if (error) {
        tagClasses.push('frui-bg-error');
    }
    else if (muted) {
        tagClasses.push('frui-bg-muted');
    }
    else {
        tagClasses.push('frui-bg-warning');
    }
    return (_jsxs("div", { ...attributes, className: classNames.join(' '), children: [tags.map((tag, i) => (_jsxs("div", { className: tagClasses.join(' '), style: tagStyle, children: [tag, _jsx("button", { className: "frui-field-taglist-remove", onClick: () => handlers.remove(i), children: "\u00D7" }), _jsx("input", { type: "hidden", name: name, value: tag })] }, i))), _jsx("input", { className: "frui-field-taglist-input", value: input, onKeyDown: handlers.edit, onKeyUp: handlers.save, onChange: handlers.change })] }));
}
