// stackpress/cjs/schema/config/typemaps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.helper = exports.sqlite = exports.pgsql = exports.mysql = exports.literal = exports.method = exports.format = exports.model = exports.type = void 0;
exports.type = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'string',
    Time: 'string',
    Datetime: 'string',
    Json: 'string',
    Object: 'string',
    Hash: 'string'
};
exports.model = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'string',
    Time: 'string',
    Datetime: 'string',
    Json: 'Record<string, string|number|boolean|null>',
    Object: 'Record<string, string|number|boolean|null>',
    Hash: 'Record<string, string|number|boolean|null>'
};
exports.format = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'string|number|Date',
    Time: 'string|number|Date',
    Datetime: 'string|number|Date',
    Json: 'Record<string, any>',
    Object: 'Record<string, any>',
    Hash: 'Record<string, any>'
};
exports.method = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'integer',
    Float: 'float',
    Boolean: 'boolean',
    Date: 'date',
    Datetime: 'date',
    Time: 'date',
    Json: 'object',
    Object: 'object',
    Hash: 'object'
};
exports.literal = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'integer',
    Float: 'float',
    Boolean: 'boolean',
    Date: 'date',
    Time: 'time',
    Datetime: 'datetime',
    Json: 'json',
    Object: 'json',
    Hash: 'json'
};
exports.mysql = {
    String: 'string',
    Text: 'text',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'date',
    Datetime: 'datetime',
    Time: 'time',
    Json: 'json',
    Object: 'json',
    Hash: 'json'
};
exports.pgsql = {
    String: 'string',
    Text: 'text',
    Number: 'number',
    Integer: 'number',
    Float: 'number',
    Boolean: 'boolean',
    Date: 'date',
    Datetime: 'timestamp',
    Time: 'time',
    Json: 'jsonb',
    Object: 'jsonb',
    Hash: 'jsonb'
};
exports.sqlite = {
    String: 'string',
    Text: 'string',
    Number: 'number',
    Integer: 'number',
    Float: 'real',
    Boolean: 'number',
    Date: 'string',
    Datetime: 'string',
    Time: 'string',
    Json: 'string',
    Object: 'string',
    Hash: 'string'
};
exports.helper = {
    String: 'toSqlString',
    Text: 'toSqlString',
    Number: 'toSqlFloat',
    Integer: 'toSqlInteger',
    Float: 'toSqlFloat',
    Boolean: 'toSqlBoolean',
    Date: 'toSqlDate',
    Time: 'toSqlDate',
    Datetime: 'toSqlDate',
    Json: 'toSqlString',
    Object: 'toSqlString',
    Hash: 'toSqlString'
};
