// stackpress/cjs/sql/sqlite.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Actions = exports.actions = exports.getAlias = exports.getColumnJoins = exports.getColumnPath = exports.getColumnInfo = exports.getColumns = exports.sequence = exports.toSqlFloat = exports.toSqlInteger = exports.toSqlDate = exports.toSqlBoolean = exports.toSqlString = exports.toResponse = exports.toErrorResponse = exports.intable = exports.boolable = exports.dateable = exports.floatable = exports.stringable = exports.connect = exports.BetterSqlite3Connection = exports.joins = exports.Exception = exports.Engine = exports.Sqlite = exports.Pgsql = exports.Mysql = exports.Update = exports.Select = exports.Insert = exports.Delete = exports.Create = exports.Alter = void 0;
const index_js_1 = __importStar(require("./actions/index.js"));
exports.actions = index_js_1.default;
Object.defineProperty(exports, "Actions", { enumerable: true, get: function () { return index_js_1.Actions; } });
var inquire_1 = require("@stackpress/inquire");
Object.defineProperty(exports, "Alter", { enumerable: true, get: function () { return inquire_1.Alter; } });
Object.defineProperty(exports, "Create", { enumerable: true, get: function () { return inquire_1.Create; } });
Object.defineProperty(exports, "Delete", { enumerable: true, get: function () { return inquire_1.Delete; } });
Object.defineProperty(exports, "Insert", { enumerable: true, get: function () { return inquire_1.Insert; } });
Object.defineProperty(exports, "Select", { enumerable: true, get: function () { return inquire_1.Select; } });
Object.defineProperty(exports, "Update", { enumerable: true, get: function () { return inquire_1.Update; } });
Object.defineProperty(exports, "Mysql", { enumerable: true, get: function () { return inquire_1.Mysql; } });
Object.defineProperty(exports, "Pgsql", { enumerable: true, get: function () { return inquire_1.Pgsql; } });
Object.defineProperty(exports, "Sqlite", { enumerable: true, get: function () { return inquire_1.Sqlite; } });
Object.defineProperty(exports, "Engine", { enumerable: true, get: function () { return inquire_1.Engine; } });
Object.defineProperty(exports, "Exception", { enumerable: true, get: function () { return inquire_1.Exception; } });
Object.defineProperty(exports, "joins", { enumerable: true, get: function () { return inquire_1.joins; } });
var inquire_sqlite3_1 = require("@stackpress/inquire-sqlite3");
Object.defineProperty(exports, "BetterSqlite3Connection", { enumerable: true, get: function () { return inquire_sqlite3_1.BetterSqlite3Connection; } });
Object.defineProperty(exports, "connect", { enumerable: true, get: function () { return inquire_sqlite3_1.connect; } });
var helpers_js_1 = require("./helpers.js");
Object.defineProperty(exports, "stringable", { enumerable: true, get: function () { return helpers_js_1.stringable; } });
Object.defineProperty(exports, "floatable", { enumerable: true, get: function () { return helpers_js_1.floatable; } });
Object.defineProperty(exports, "dateable", { enumerable: true, get: function () { return helpers_js_1.dateable; } });
Object.defineProperty(exports, "boolable", { enumerable: true, get: function () { return helpers_js_1.boolable; } });
Object.defineProperty(exports, "intable", { enumerable: true, get: function () { return helpers_js_1.intable; } });
Object.defineProperty(exports, "toErrorResponse", { enumerable: true, get: function () { return helpers_js_1.toErrorResponse; } });
Object.defineProperty(exports, "toResponse", { enumerable: true, get: function () { return helpers_js_1.toResponse; } });
Object.defineProperty(exports, "toSqlString", { enumerable: true, get: function () { return helpers_js_1.toSqlString; } });
Object.defineProperty(exports, "toSqlBoolean", { enumerable: true, get: function () { return helpers_js_1.toSqlBoolean; } });
Object.defineProperty(exports, "toSqlDate", { enumerable: true, get: function () { return helpers_js_1.toSqlDate; } });
Object.defineProperty(exports, "toSqlInteger", { enumerable: true, get: function () { return helpers_js_1.toSqlInteger; } });
Object.defineProperty(exports, "toSqlFloat", { enumerable: true, get: function () { return helpers_js_1.toSqlFloat; } });
Object.defineProperty(exports, "sequence", { enumerable: true, get: function () { return helpers_js_1.sequence; } });
Object.defineProperty(exports, "getColumns", { enumerable: true, get: function () { return helpers_js_1.getColumns; } });
Object.defineProperty(exports, "getColumnInfo", { enumerable: true, get: function () { return helpers_js_1.getColumnInfo; } });
Object.defineProperty(exports, "getColumnPath", { enumerable: true, get: function () { return helpers_js_1.getColumnPath; } });
Object.defineProperty(exports, "getColumnJoins", { enumerable: true, get: function () { return helpers_js_1.getColumnJoins; } });
Object.defineProperty(exports, "getAlias", { enumerable: true, get: function () { return helpers_js_1.getAlias; } });
