// stackpress/cjs/sql/events/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.upsert = exports.update = exports.search = exports.restore = exports.remove = exports.purge = exports.get = exports.detail = exports.create = exports.batch = void 0;
exports.handlers = handlers;
exports.default = listen;
const Server_1 = require("@stackpress/ingest/Server");
const batch_js_1 = __importDefault(require("./batch.js"));
exports.batch = batch_js_1.default;
const create_js_1 = __importDefault(require("./create.js"));
exports.create = create_js_1.default;
const detail_js_1 = __importDefault(require("./detail.js"));
exports.detail = detail_js_1.default;
const get_js_1 = __importDefault(require("./get.js"));
exports.get = get_js_1.default;
const purge_js_1 = __importDefault(require("./purge.js"));
exports.purge = purge_js_1.default;
const remove_js_1 = __importDefault(require("./remove.js"));
exports.remove = remove_js_1.default;
const restore_js_1 = __importDefault(require("./restore.js"));
exports.restore = restore_js_1.default;
const search_js_1 = __importDefault(require("./search.js"));
exports.search = search_js_1.default;
const update_js_1 = __importDefault(require("./update.js"));
exports.update = update_js_1.default;
const upsert_js_1 = __importDefault(require("./upsert.js"));
exports.upsert = upsert_js_1.default;
function handlers(model) {
    return {
        batch: (0, batch_js_1.default)(model),
        create: (0, create_js_1.default)(model),
        detail: (0, detail_js_1.default)(model),
        get: (0, get_js_1.default)(model),
        purge: (0, purge_js_1.default)(model),
        remove: (0, remove_js_1.default)(model),
        restore: (0, restore_js_1.default)(model),
        search: (0, search_js_1.default)(model),
        update: (0, update_js_1.default)(model),
        upsert: (0, upsert_js_1.default)(model)
    };
}
function listen(model) {
    const emitter = (0, Server_1.server)();
    emitter.on(`${model.dash}-batch`, (0, batch_js_1.default)(model));
    emitter.on(`${model.dash}-create`, (0, create_js_1.default)(model));
    emitter.on(`${model.dash}-detail`, (0, detail_js_1.default)(model));
    emitter.on(`${model.dash}-get`, (0, get_js_1.default)(model));
    emitter.on(`${model.dash}-purge`, (0, purge_js_1.default)(model));
    emitter.on(`${model.dash}-remove`, (0, remove_js_1.default)(model));
    emitter.on(`${model.dash}-restore`, (0, restore_js_1.default)(model));
    emitter.on(`${model.dash}-search`, (0, search_js_1.default)(model));
    emitter.on(`${model.dash}-update`, (0, update_js_1.default)(model));
    emitter.on(`${model.dash}-upsert`, (0, upsert_js_1.default)(model));
    return emitter;
}
;
